---
title: "Analysis"
author: "Alexis Hill"
date: "09/23/2023"
output: html_document
---

### Libraries
```{r, echo = FALSE, message = FALSE, warning=FALSE}
require(ggplot2)
require(tidyverse)
require(cowplot)
require(scales)
require(kableExtra)
require(RColorBrewer)
```

#### Summary
Fit Logel proposed model (new prom & terminator in second gene grouping) to high-resolution RNA-seq data

Common Functions & Base directory 
```{r, echo = FALSE, message = FALSE, warning=FALSE}
# Functions & Variables I use a lot
date <- "10012023"
filedir <- paste("../../../../output/", date, "/", sep="")

my_palette = c('#8C4C47', '#C27F7A', '#21819F', '#65AFCF', '#93D8F8', '#398048', '#73B47E', '#97D8A2', '#A56EA9', '#C994CE', '#EDB7F2')

# get_reports() retrieves the reports for all the simulation runs in a specified scenario
# had hardcoded case for Scenario B where there is no simulation 1
get_reports <- function(filedir){
  report <- data.frame()
   batch <- 1:20

  
  for(i in batch){
     sim = read_csv(paste0(filedir, "", "sim_", i, "_report.csv")) 
     sim$sim = i
     report <- rbind(report, sim)}
  return(report)
}

get_single_report <- function(filedir, sim){
  return(read_csv(paste0(filedir, "", "sim_", sim, "_report.csv"))) 
}

best_gen <- function(filedir) {
  get_reports(filedir) %>% 
  group_by(sim) %>% 
  mutate(final_min_error = min(min_error)) %>% 
  filter(error == final_min_error) %>%  
  filter(row_number()==1) %>% 
  #filter(min_error < 2) %>% 
  arrange(min_error)
}

# plot_reports() plots the RMSE v generations for all the simulations in a given scenario
plot_reports <- function(filedir,  title){
  x_scale <- c(0, 100, 200, 300, 400, 500)
  
  figure <- get_reports(filedir) %>% 
    # filter(!is.na(gen)) %>% 
    filter(gen > 0) %>% 
    filter(error < 10000) %>% 
    ggplot(aes(x = gen)) +
    geom_line(aes(y = error, color = "error"), color = "#CC79A7") + 
    geom_line(aes(y = min_error, color = "min error"), color = "black") +
    facet_wrap(~sim) + 
    #ggtitle(title) + 
    xlab("Generations") +
    ylab("RMSE") + 
    #ylim(c(1, 10)) + 
    #scale_x_continuous(breaks = x_scale) + 
    cowplot::theme_minimal_grid(12) + 
    theme(axis.text.x = element_text(size=11, angle= -45))
  
  return(figure)
}

logel_plot <- function(filedir, sim, gen, timestamp){

  figure <- read_tsv(paste0(filedir,"sim_", sim, "_gen_", gen, "_ID_0", ".tsv" )) %>% 
       mutate(time = round(time)) %>% 
       filter(time == timestamp) %>% 
       filter(!str_detect(species, '__|bound|ecoli|promoter|degraded')) %>% 
       #mutate(normalized = transcript/transcript[1]) %>% 
       mutate(species = factor(species, c("gene_A", "gene_A*", "gene_B",
                                         "gene_K", "gene_C", "gene_D",
                                         "gene_E", "gene_J", "gene_F",
                                         "gene_G", "gene_H"),
                                       c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"))) %>% 
       ggplot(aes(x=species, y=transcript, fill = species)) +
       geom_bar(stat="identity", color = "black", alpha = 0.75) + 
       scale_fill_manual(values = my_palette) + 
       #scale_y_continuous() + 
       #scale_y_continuous(limits = c(0, 18),
        #                  breaks = c(0,3,6,9,12,15,18)) +
       scale_y_continuous(limits = c(0, 300), expand = c(0,0)) +
       ylab("")+
       cowplot::theme_cowplot(12) +
       theme(legend.position = "none") 
  
  return(figure)
}

logel_plot_2 <- function(df){
  
  figure <- df %>% 
       filter(!str_detect(species, '__|bound|ecoli|promoter|degraded')) %>% 
       #mutate(normalized = transcript/transcript[1]) %>% 
       mutate(species = factor(species, c("gene_A", "gene_A*", "gene_B",
                                         "gene_K", "gene_C", "gene_D",
                                         "gene_E", "gene_J", "gene_F",
                                         "gene_G", "gene_H"),
                                       c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"))) %>% 
       ggplot(aes(x=species, y=transcript, fill = species)) +
       geom_bar(stat="identity", color = "black") + 
       scale_fill_manual(values = my_palette) + 
       #scale_y_continuous() + 
       #scale_y_continuous(limits = c(0, 18),
        #                  breaks = c(0,3,6,9,12,15,18)) +
       scale_y_continuous(limits = c(0, 300), expand = c(0,0)) +
       ylab("") +
       xlab("genes") + 
       cowplot::theme_cowplot(12) +
       theme(legend.position = "none") 
  
  return(figure)
}

logel_plot_prot <- function(filedir, sim, gen, timestamp){

  figure <- read_tsv(paste0(filedir,"sim_", sim, "_gen_", gen, "_ID_0", ".tsv" )) %>% 
       mutate(time = round(time)) %>% 
       filter(time == timestamp) %>% 
       filter(!str_detect(species, '__|bound|ecoli|promoter|degraded')) %>% 
       #mutate(normalized = transcript/transcript[1]) %>% 
       mutate(species = factor(species, c("gene_A", "gene_A*", "gene_B",
                                         "gene_K", "gene_C", "gene_D",
                                         "gene_E", "gene_J", "gene_F",
                                         "gene_G", "gene_H"),
                                       c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"))) %>% 
       ggplot(aes(x=species, y=protein, fill = species)) +
       geom_bar(stat="identity", color = "black") + 
       scale_fill_manual(values = my_palette) + 
       #scale_y_continuous() + 
       #scale_y_continuous(limits = c(0, 18),
        #                  breaks = c(0,3,6,9,12,15,18)) +
       scale_y_continuous(limits = c(0, 15000), expand = c(0,0)) +
       ylab("Simulated protein")+
       xlab("Genes")+ 
       cowplot::theme_cowplot(12) +
       theme(legend.position = "none") 
  
  return(figure)
}

get_fitness <- function(file, timepoint){
  
  pt_run <- as.data.frame(read_tsv(file)) %>% 
    mutate(time = round(time, 0)) %>% 
    filter(time == timepoint) %>% 
    filter(!str_detect(species, '__|bound|ecoli|promoter')) %>% 
    mutate(species = factor(species, c("gene_A", "gene_A*", "gene_B",
                                           "gene_K", "gene_C", "gene_D",
                                           "gene_E", "gene_J", "gene_F",
                                           "gene_G", "gene_H"),
                                     c("A", "A*", "B", "K",
                                           "C", "D", "E", "J",
                                           "F", "G", "H"))) %>% 
    filter(species %in% c("B", "D", "J", "F", "G", "H")) %>% 
    select(-ribo_density, -transcript, -time) %>% 
    mutate(copies = c(60, 240, 60, 60, 12, 60))  %>% # From Leuven 2020
    mutate(stoich = protein/copies) %>%
    summarise(virions = min(stoich))
  
  return(pt_run$virions[1])
}

logel_timeseries_with_pcr <- function(filedir, sim, gen1, gen2, gen3, timestamp) {
  
  sim_qpcr <- data.frame(species = c("A", "A*", "B", "K","C", "D", "E", "J","F", "G", "H"),
                        normalized = c(17, 17, 102, 102, 102, 289, 289, 289, 187, 85, 17)) %>% 
            mutate(species = factor(species, c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"),
                                       c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"))) %>% 
            ggplot(aes(x=species, y=normalized, fill = species)) +
            geom_bar(stat="identity", color = "black") + 
            scale_fill_manual(values = my_palette) + 
            scale_y_continuous(limits = c(0, 300)) + 
            #breaks = c(0,3,6,9,12,15,18)) + 
            ggtitle("Target (qPCR)") + 
            ylab("Relative transcripts")+ 
            xlab("Î¦X174 genes")+ 
            cowplot::theme_cowplot(12) +
            theme(legend.position = "none",
                  plot.title = element_text(hjust = 0.5)) 
  
  sim_start <- logel_plot(filedir, sim, gen = gen1, timestamp = timestamp) + 
               ggtitle(paste0("Generation ", gen1)) + 
               theme(plot.title = element_text(hjust = 0.5))
  sim_middle <- logel_plot(filedir, sim, gen = gen2, timestamp = timestamp) + 
                ggtitle(paste0("Generation ", gen2)) +
                theme(plot.title = element_text(hjust = 0.5))
  sim_end <- logel_plot(filedir, sim, gen = gen3, timestamp = timestamp) + 
               ggtitle(paste0("Final", "")) + 
               theme(plot.title = element_text(hjust = 0.5))
  
  #title <- ggdraw() + draw_label(paste0("", "Simulated Transcription Patterns"))
  
  figure <- cowplot::plot_grid(sim_qpcr, sim_start, sim_middle, sim_end,
                      #label_x = "genes", label_y = "transcript abundances", 
                      #labels = c("A", "B", "", ""), 
                      ncol = 4, nrow = 1)
  
  #figure <- plot_grid(panels, ncol=1, rel_heights=c(0.1, 1)) 
               
  return(figure)
}

logel_timeseries_with_rnaseq <- function(filedir, sim, gen1, gen2, gen3, timestamp) {
  
  sim_qpcr <- data.frame(species = c("A", "A*", "B", "K","C", "D", "E", "J","F", "G", "H"),
                        normalized = c(27, 27, 120, 81, 123, 228, 201, 264, 231, 118, 76)) %>% 
            mutate(species = factor(species, c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"),
                                       c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"))) %>% 
            ggplot(aes(x=species, y=normalized, fill = species)) +
            geom_bar(stat="identity", color = "black") + 
            scale_fill_manual(values = my_palette) + 
            scale_y_continuous(limits = c(0, 300), expand = c(0,0)) + 
            #breaks = c(0,3,6,9,12,15,18)) + 
            ggtitle("Target (RNAseq)") + 
            ylab("Transcripts")+ 
            xlab("Î¦X174 genes")+ 
            cowplot::theme_cowplot(12) +
            theme(legend.position = "none",
                  plot.title = element_text(hjust = 0.5)) 
  
  sim_start <- logel_plot(filedir, sim, gen = gen1, timestamp = timestamp) + 
               ggtitle(paste0("Generation ", gen1)) + 
               theme(plot.title = element_text(hjust = 0.5))
  sim_middle <- logel_plot(filedir, sim, gen = gen2, timestamp = timestamp) + 
                ggtitle(paste0("Generation ", gen2)) +
                theme(plot.title = element_text(hjust = 0.5))
  sim_end <- logel_plot(filedir, sim, gen = gen3, timestamp = timestamp) + 
               ggtitle(paste0("Final", "")) + 
               theme(plot.title = element_text(hjust = 0.5))
  
  #title <- ggdraw() + draw_label(paste0("", "Simulated Transcription Patterns"))
  
  figure <- cowplot::plot_grid(sim_qpcr, sim_start, sim_middle, sim_end,
                      #label_x = "genes", label_y = "transcript abundances", 
                      #labels = c("A", "B", "", ""), 
                      ncol = 4, nrow = 1)
  
  #figure <- plot_grid(panels, ncol=1, rel_heights=c(0.1, 1)) 
               
  return(figure)
}

timeseries_proteomics <- function(filedir, sim, gen1, gen2, gen3, timestamp) {
  
  sim_wt <- data.frame(species = c("A", "A*", "B", "K","C", "D", "E", "J","F", "G", "H"),
            normalized = c(669.8, 573.4, 4266, 3630, 3450, 12403, 10480, 10219, 6826, 2875, 538)) %>%
            mutate(species = factor(species, c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"),
                                       c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"))) %>% 
            ggplot(aes(x=species, y=normalized, fill = species)) +
            geom_bar(stat="identity", color = "black") + 
            scale_fill_manual(values = my_palette) + 
            scale_y_continuous(limits = c(0, 15000), expand = c(0,0)) + 
            #breaks = c(0,3,6,9,12,15,18)) + 
            ggtitle(expression("WT")) + 
            ylab("Simulated protein")+ 
            xlab("Genes")+ 
            cowplot::theme_cowplot(12) +
            theme(legend.position = "none",
                  plot.title = element_text(hjust = 0.5)) 
  
  sim_decomp <- data.frame(species = c("A", "A*", "B", "K","C", "D", "E", "J","F", "G", "H"),
            normalized = c(6698, 5734, 711, 605, 575, 12403, 10480, 10219, 6826, 2875, 538)) %>%
            mutate(species = factor(species, c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"),
                                       c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"))) %>% 
            ggplot(aes(x=species, y=normalized, fill = species)) +
            geom_bar(stat="identity", color = "black") + 
            scale_fill_manual(values = my_palette) + 
            scale_y_continuous(limits = c(0, 15000), expand = c(0,0)) + 
            #breaks = c(0,3,6,9,12,15,18)) + 
            ggtitle(expression("Decompressed (pred.)")) + 
            ylab("Simulated protein")+ 
            xlab("Genes")+ 
            cowplot::theme_cowplot(12) +
            theme(legend.position = "none",
                  plot.title = element_text(hjust = 0.5)) 
  
  sim_end <- logel_plot_prot(filedir, sim, gen = gen3, timestamp = timestamp) + 
               ggtitle(expression("Decompressed (fit)")) + 
               theme(plot.title = element_text(hjust = 0.5))
  

  figure <- cowplot::plot_grid(sim_wt, sim_decomp, sim_end,
                      #label_x = "genes", label_y = "transcript abundances", 
                      #labels = c("A", "B", "", ""), 
                      ncol = 3, nrow = 1)
  

  return(figure)
}

exp_rnaseq <- function() {
  figure <- data.frame(species = c("A", "A*", "B", "K","C", "D", "E", "J","F", "G", "H"),
                        normalized = c(27, 27, 120, 81, 123, 228, 201, 264, 231, 118, 76)) %>% 
            mutate(species = factor(species, c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"),
                                       c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"))) %>% 
            ggplot(aes(x=species, y=normalized, fill = species)) +
            geom_bar(stat="identity", color = "black") + 
            scale_fill_manual(values = my_palette) + 
            scale_y_continuous(limits = c(0, 300)) + 
            #breaks = c(0,3,6,9,12,15,18)) + 
            ggtitle("Target (RNA-seq)") + 
            ylab("Transcripts")+ 
            xlab("genes")+ 
            cowplot::theme_cowplot(12) +
            theme(legend.position = "none",
                  plot.title = element_text(hjust = 0.5)) 
  return(figure)
}

exp_qpcr <- function() {
  figure <- data.frame(species = c("A", "A*", "B", "K","C", "D", "E", "J","F", "G", "H"),
                        normalized = c(17, 17, 102, 102, 102, 289, 289, 289, 187, 85, 17)) %>% 
            mutate(species = factor(species, c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"),
                                       c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"))) %>% 
            ggplot(aes(x=species, y=normalized, fill = species)) +
            geom_bar(stat="identity", color = "black") + 
            scale_fill_manual(values = my_palette) + 
            scale_y_continuous(limits = c(0, 300)) + 
            #breaks = c(0,3,6,9,12,15,18)) + 
            ggtitle("Target (qPCR)") + 
            ylab("Transcripts")+ 
            xlab("genes")+ 
            cowplot::theme_cowplot(12) +
            theme(legend.position = "none",
                  plot.title = element_text(hjust = 0.5)) 
  return(figure)
}

transcripts_vs_time_plot <- function(filedir, sim, gen){

  figure <- read_tsv(paste0(filedir,"sim_", sim, "_gen_", gen, "_ID_0", ".tsv" )) %>% 
       filter(!str_detect(species, '__|bound|ecoli|promoter|degraded')) %>% 
       mutate(species = factor(species, c("gene_A", "gene_A*", "gene_B",
                                         "gene_K", "gene_C", "gene_D",
                                         "gene_E", "gene_J", "gene_F",
                                         "gene_G", "gene_H"),
                                       c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"))) %>% 
       ggplot(aes(x=time, y=transcript, color = species)) +
       geom_line() + 
       scale_color_manual(values = my_palette) + 
       ylab("Transcript counts")+
       cowplot::theme_cowplot(12)
  
  return(figure)
}

```

```{r}
plot_reports(filedir, "Decompressed")
```

```{r}
best_gen_report <- best_gen(filedir)
best_gen_report
```

```{r}
decomp_sims <- timeseries_proteomics(filedir, "13", "0", "500", "3423", 500)
decomp_sims
```
```{r}
promA <- expression(italic("pA")); promB <- expression(italic("pB")); promD <- expression(italic("pD"))
tJ <- expression("T"["J"]); tF <- expression("T"["F"]); tG <- expression("T"["G"]); tH <- expression("T"["H"]);

promoter_dist_2 <- best_gen_report %>% 
  #filter(final_min_error <= 2.8) %>%
  select(-1:-2, -10:-13) %>%
  pivot_longer(!sim, names_to = "element", values_to = "strength") %>%
  filter(element %in% c("pA", "pB", "pD")) %>%
  ggplot(aes(x=fct_relevel(element, "pA", "pB", "pD"), y=strength)) +
  geom_boxplot(fill = "gray67") +
  geom_point(data = data.frame(x = c("pA","pB","pD"), y = c(11.32, 3.34, 34.37)),
             aes(x=x, y=y),
             color = 'lightblue', fill = 'lightblue',
             shape = 23, size = 2) +
  scale_y_continuous(limits = c(0, 40), expand = expansion(mult = c(0, 0))) +
  scale_fill_manual(values = prom_palette) +
  #scale_x_discrete(labels = c(promA, promB, promD)) +
  ylab("Binding strength") +
  #ylab(expression("binding strength" ~ 
  #                  (10^{6} ~ M^{-1} ~ s^{-1})
  #)) +
  xlab(NULL) +
  theme_minimal() +
  panel_border(color = "gray40") +
  theme(legend.position = "none",
        panel.grid.minor.x = element_blank())

terminator_dist_2 <- best_gen_report %>% 
  #filter(final_min_error <= 2.8) %>%
  select(-1:-2, -10:-13) %>%
  pivot_longer(!sim, names_to = "element", values_to = "strength") %>%
  filter(!element %in% c("pA", "pB", "pD")) %>%
  ggplot(aes(x=fct_relevel(element, "tJ", "tF", "tG", "tH"), y=strength)) +
  geom_boxplot(fill = "gray67") +
  geom_point(data = data.frame(x = c("tJ","tF","tG", "tH"), y = c(0.34, 0.51, 0.78, 0.93)),
             aes(x=x, y=y),
             color = 'lightblue', fill = 'lightblue',
             shape = 23, size = 2) +
  scale_y_continuous(limits = c(0, 1), expand = expansion(mult = c(0, 0))) +
  scale_fill_manual(values = term_palette) +
  scale_x_discrete(labels = c(tJ, tF, tG, tH)) +
  ylab("Termination efficiency") +
  xlab(NULL) +
  theme_minimal() +
  panel_border(color = "gray40") +
  theme(legend.position = "none",
        panel.grid.minor.x = element_blank())

decomp_dists <- cowplot::plot_grid(promoter_dist_2, terminator_dist_2,
                  #label_x = "genes", label_y = "transcript abundances", 
                  labels = c("A", "B"), 
                  ncol = 2, nrow = 1, align = "h")

cowplot::plot_grid(decomp_dists, decomp_sims, labels = c("", "C"), nrow = 2)
ggsave("paper_figures/decomp_S5.png", width = 9, height = 5)
```


