---
title: "Figures"
author: "Tanvi Ingle"
date: "7/13/2022"
output: html_document
---
### **Objectives**
This Rmd file contains code snippets that recreate figures found in my undergrad thesis. The original code to recreate the figures are found in Wilke/tingle/phix174/src/r/thesis_figures/Thesis_Figures.Rmd.  Important vocabulary include: 

**simulation** = the optimization attempt (ie, 1 run of the phix174_opt_12.py script)
**generation** = an attempt within a simulation run to decrease RMSE
**timestamp** = timepoint in the infection cycle for PhiX174. 

Essentially, each simulation run will have 500/1000 generations. Each generation is a run of the PhiX174 pinetree model for one infection cycle (as set by timestamp).

Directory of Each Figure from UG thesis:

1. **Figure 1 & 2** were created in powerpoint. It was repurposed from Wright et. al.; not sure if this will have to be redone in a new format to avoid copyright issues. 

2. **Figure 3** from `plot_multi_rmse_curves()`

3. **Figure 4** from `logel_timeseries()`

4. **Figure 5** from `compare_wt_dc_transcripts_plot()`

5. **Figure 6** from `compare_wt_dc_absolute_transcripts_plot()`

6. **Figure 7** from `plot_fitness()`

#### **Set Libraries and Working Directories**
Update file pathways. Defaults set to directores in the `tingle/` directory on the Wilke pods. 

`basedir`: working directory
`filedir`: folder in `basedir` that contains simulation outputs and reports
`figuredir`: folder in `basedir` that contains saved figures created in this Rmd

```{r, echo = FALSE, message = FALSE, warning=FALSE}
require(ggplot2)
require(tidyverse)
require(cowplot)
require(scales)
require(kableExtra)
library("docxtools")

basedir <- "/stor/work/Wilke/tingle/phix174/src/python/pycharm/new_scripts/"
filedir <- "output/"
figuredir <- paste0(basedir, "figures/")
```



### **Functions**
#### `get_reports()`: 

**FUNCTION** - returns a dataframe with all of the reports saved from each simulation run. Each report stores data on the trial number, calculated RMSE, and minimum overall RMSE values. 

**ARGUMENTS** - basedir, filedir, and total_sims (the total number of simulations, or attempted optimizations, run)

```{r}
get_reports <- function(basedir, filedir, total_sims){
  report <- data.frame()
  i = 1
  while (i<=total_sims) {
    sim = read_csv(paste0(basedir, filedir, "sim_", i, "_report.csv")) 
    sim$sim = i
    report <- rbind(report, sim)
    i <- i + 1
  }
  return(report)
}
```


#### `plot_reports()`: 

**FUNCTION** - creates a figure that plots all of the RMSE values & the lowest RMSE value, faceted for each simulation run. 

**ARGUMENTS** - basedir, filedir, total_sims, start_sim (the first simulation run you want to include in the plot), end_sim (the last simulation run to include in the plot), title. 

```{r}
plot_reports <- function(basedir, filedir, total_sims, start_sim, end_sim, title){
  x_scale <- c(0, 100, 200, 300, 400, 500)
  y_scale <- c(0, 3, 6, 9, 12)
  mycolors <- c("black", "#CC79A7")
  
  figure <- get_reports(basedir, filedir, total_sims) %>% 
    filter(sim %in% c(start_sim : end_sim)) %>% 
    filter(!is.na(gen)) %>% 
    filter(gen > 0) %>% 
    filter(error <= 12) %>% 
    ggplot(aes(x = gen)) +
    # geom_line(aes(y = error, color = "error"), color = "#CC79A7") + 
    # geom_line(aes(y = min_error, color = "min_error"), color = "black") +
    geom_line(aes(y = error, color = "RMSE")) +
    geom_line(aes(y = min_error, color = "Lowest RMSE")) +
    labs(colour="") +
    facet_wrap(~sim) + 
    ggtitle(title) + 
    xlab("Generations") +
    ylab("RMSE") + 
    scale_y_continuous(breaks = y_scale) + 
    scale_x_continuous(breaks = x_scale) + 
    scale_color_manual(values = mycolors) + 
    cowplot::theme_minimal_grid(12) + 
    theme(axis.text.x = element_text(size=12, angle= -45))
  
  return(figure)
}
```



#### `logel_plot()`: 

**FUNCTION** - creates a figure similar to the one in Logel 2020 which plots the transcript abundances for each of the PhiX-174 genes relative to the expression levels of gene A. 

**ARGUMENTS** - basedir, filedir, sim (the specific simulation number to be plotted), gen (the generation, ie optimization attempt in a single simulation run, to be plotted), timestamp (the timepoint in the simulation run to be plotted. generally this is at the end of the infection cycle, so 1200).

**WARNING** - the colors do NOT exactly match the figure in Logel 2020. @TODO need to fix this. 

*Commented out y_scales in logel(); uncomment this before running the timeseries plot.*

```{r}
logel_plot <- function(basedir, filedir, sim, gen, timestamp){
  
  my_palette = c("#d0a34f","#75c130", "#c12c04", "#ecd575", "#b69507",
            "#e37c71", "#adcb40", "#d3ad40", "#288ede", "#a5bdfa",
            "#93c9ff")

  figure <- read_tsv(paste0(basedir, filedir,"sim_", sim, "_gen_", gen, ".tsv" )) %>% 
       mutate(time = round(time)) %>% 
       filter(time == timestamp) %>% 
       filter(!str_detect(species, '__|bound|ecoli|promoter|degraded')) %>% 
       mutate(normalized = transcript/transcript[1]) %>% 
       mutate(species = factor(species, c("gene_A", "gene_A*", "gene_B",
                                         "gene_K", "gene_C", "gene_D",
                                         "gene_E", "gene_J", "gene_F",
                                         "gene_G", "gene_H"),
                                       c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"))) %>% 
       ggplot(aes(x=species, y=normalized, fill = species)) +
       geom_bar(stat="identity", color = "black") + 
       scale_fill_manual(values = my_palette) + 
       #scale_y_continuous() + 
       # scale_y_continuous(limits = c(0, 19),
       #                    breaks = c(0,3,6,9,12,15,18)) +
       ylab("")+
       xlab("ΦX174 genes") + 
       cowplot::theme_cowplot(12) +
       theme(legend.position = "none") 
  
  return(figure)
}
```


#### `rmse_curves()`: 

**FUNCTION** - plots the RMSE and minimum RMSE curves for 1 simulation. 

**ARGUMENTS** - basedir, filedir, sim, timestamp. 

**WARNING** - RMSE values greater than 12 are exclused from the plot.  

```{r}
rmse_curves <- function(basedir, filedir, sim, timestamp){
 
  x_scale <- c(0, 100, 200, 300, 400, 500)
  y_scale <- c(0, 3, 6, 9, 12)
  mycolors <- c("black", "#CC79A7")
  
  figure <- read_csv(paste0(basedir, "sim_", sim, "_report.csv")) %>% 
            filter(!is.na(gen)) %>% 
            filter(gen > 0) %>% 
            filter(error <= 12) %>% 
            ggplot(aes(x = gen)) +
            geom_line(aes(y = error, color = "RMSE")) +
            geom_line(aes(y = min_error, color = "Lowest RMSE")) +
            labs(colour="") +
            ggtitle(paste0("Trial ", sim)) + 
            xlab("Generation") +
            ylab("RMSE") + 
            scale_y_continuous(breaks = y_scale) + 
            scale_x_continuous(breaks = x_scale) + 
            scale_color_manual(values = mycolors) + 
            cowplot::theme_minimal_grid(12) + 
            theme(axis.text.x = element_text(size=12, angle= -45),
                  plot.title = element_text(hjust = 0.5))
  
  return(figure)
}
```
#### `logel_timeseries()`

**FUNCTION** - creates a 4 column, 1 row plot of transcript abundances for 1 simulation and three different generations (ex Simulation 10, generations 1, 250, and 500). The first column in the plot is the experimental qPCR relative abundances, described in Logel 2020. Plots in columns 2, 3, and 4 correspond to first, second, and third generations (respectively) specified by the user. The purpose of this plot is to showcase that as RMSE values drop, the overall transcript abundance pattern evolves to be closer to qPCR data. 

**ARGUMENTS** - basedir, filedir, sim, gen1, gen2, gen3 (each gen reflects 3 different genreation timepoints), timestamp (point in the infection cycle, typically 1200).

```{r}
logel_timeseries <- function(basedir, filedir, sim, gen1, gen2, gen3, timestamp) {
  
 
  my_palette = c("#d0a34f","#75c130", "#c12c04", "#ecd575", "#b69507",
            "#e37c71", "#adcb40", "#d3ad40", "#288ede", "#a5bdfa",
            "#93c9ff")
  
  sim_qpcr <- data.frame(species = c("A", "A*", "B", "K","C", "D", "E", "J","F", "G", "H"),
                        normalized = c(1,1,6,6,6,17,17,17,11,5,1)) %>% 
            mutate(species = factor(species, c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"),
                                       c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"))) %>% 
            ggplot(aes(x=species, y=normalized, fill = species)) +
            geom_bar(stat="identity", color = "black") + 
            scale_fill_manual(values = my_palette) + 
            scale_y_continuous(limits = c(0, 18), 
            breaks = c(0,3,6,9,12,15,18)) + 
            ggtitle("qPCR Data") + 
            ylab("Transcript Abundances")+ 
            xlab("ΦX174 genes")+ 
            cowplot::theme_cowplot(12) +
            theme(legend.position = "none",
                  plot.title = element_text(hjust = 0.5)) 
  
  sim_start <- logel_plot(basedir, filedir, sim, gen = gen1, timestamp = 1200) + 
               ggtitle(paste0("Generation ", gen1)) + 
               theme(plot.title = element_text(hjust = 0.5))
  sim_middle <- logel_plot(basedir, filedir, sim, gen = gen2, timestamp = 1200) + 
                ggtitle(paste0("Generation ", gen2)) +
                theme(plot.title = element_text(hjust = 0.5))
  sim_end <- logel_plot(basedir, filedir, sim, gen = gen3, timestamp = 1200) + 
               ggtitle(paste0("Generation ", gen3)) + 
               theme(plot.title = element_text(hjust = 0.5))
  
  title <- ggdraw() + draw_label(paste0("Trial ", sim, ": Simulated Transcription Patterns"), fontface='bold')
  
  panels <- cowplot::plot_grid(sim_qpcr, sim_start, sim_middle, sim_end,
                      #label_x = "genes", label_y = "transcript abundances", 
                      labels = c("A", "B", "", ""), 
                      ncol = 4, nrow = 1)
  
  figure <- plot_grid(title, panels, ncol=1, rel_heights=c(0.1, 1)) 
               
  return(figure)
}
```


#### `get_fitness()`: 

**FUNCTION** - calculate the total number of virions that could be assembled given the protein abundances found in a pinetree run (file = sim_#_gen_#.tsv, timepoint = 1200). The stoichiometric relationship between protein abundances is found in Leuven 2020. 

**ARGUMENTS** - basedir, filedir, sim, gen, timestamp. 
```{r}
get_fitness <- function(basedir, filedir, sim, gen, timestamp){
  
  file <- paste0(basedir, filedir,"sim_", sim, "_gen_", gen, ".tsv")
  
  pt_run <- as.data.frame(read_tsv(file)) %>% 
    mutate(time = round(time, 0)) %>% 
    filter(time == timestamp) %>% 
    filter(!str_detect(species, '__|bound|ecoli|promoter')) %>% 
    mutate(species = factor(species, c("gene_A", "gene_A*", "gene_B",
                                           "gene_K", "gene_C", "gene_D",
                                           "gene_E", "gene_J", "gene_F",
                                           "gene_G", "gene_H"),
                                     c("A", "A*", "B", "K",
                                           "C", "D", "E", "J",
                                           "F", "G", "H"))) %>% 
    filter(species %in% c("B", "D", "J", "F", "G", "H")) %>% 
    select(-ribo_density, -transcript, -time) %>% 
    mutate(copies = c(60, 240, 60, 60, 12, 60))  %>% # From Leuven 2020
    mutate(stoich = protein/copies) %>%
    summarise(virions = min(stoich))
  
  return(pt_run$virions[1])
}
```

#### `absolute_logel()`: 

**FUNCTION** - plot the total transcript raw counts for a specific simulation and generation (NOT relative to gene A expression levels). 

**ARGUMENTS** - basedir, filedir, sim, gen, timestamp. 
```{r}
absolute_logel <- function(basedir, filedir, sim, gen, timestamp){
  
  my_palette = c("#d0a34f","#75c130", "#c12c04", "#ecd575", "#b69507",
            "#e37c71", "#adcb40", "#d3ad40", "#288ede", "#a5bdfa",
            "#93c9ff")

  figure <- read_tsv(paste0(basedir, filedir,"sim_", sim, "_gen_", gen, ".tsv" )) %>% 
       mutate(time = round(time)) %>% 
       filter(time == timestamp) %>% 
       filter(!str_detect(species, '__|bound|ecoli|promoter|degraded')) %>% 
       #mutate(normalized = transcript/transcript[1]) %>% 
       mutate(species = factor(species, c("gene_A", "gene_A*", "gene_B",
                                         "gene_K", "gene_C", "gene_D",
                                         "gene_E", "gene_J", "gene_F",
                                         "gene_G", "gene_H"),
                                       c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"))) %>% 
       ggplot(aes(x=species, y=transcript, fill = species)) +
       geom_bar(stat="identity", color = "black") + 
       scale_fill_manual(values = my_palette) + 
       #scale_y_continuous() + 
       # scale_y_continuous(limits = c(0, 19),
       #                    breaks = c(0,3,6,9,12,15,18)) +
       ylab("")+
       xlab("ΦX174 genes") + 
       cowplot::theme_cowplot(12) +
       theme(legend.position = "none") 
  
  return(figure)
}
```





### **Other Code Snippets**
#### **IMPORTANT TO READ** 

DON'T hit run on these code blocks before reading the following descriptions -- they require some finagling before they will produce the figures you want. These are snippets borrowed from the code I used to make the figures in my thesis. I did not write pretty wrapper functions for these since I wanted to provide some flexbility with the stylizing (ie I got a wee bit lazy. As my penance, I have included descriptions of how to get these blocks to run & how to customize it with the latest output data). You will notice a lot of redundancy ... but I think it's a small price to pay for improved readability. 

**STEP 1** If you want to see the figures that these code blocks produce, uncomment the following code. This will redirect the basedir and filedir to output produced for my thesis. 

`basedir` <- filepath to 4/15/2022 simulation output files. Has all the .tsv files produced during the optimization script
`filedir` <- filepath to folder within 4/15/2022 which has runs for the decompressed genome. 

```{r}
# basedir <- "/stor/work/Wilke/tingle/phix174/output/4152022/"
#  filedir <- "decompressed_runs/" 
```

**STEP 2** 
These functions are used to create figures from my thesis. They require some more manual manipulation of some hard coded elements - so be sure to read all of the documentation that proceeds each function before implementing it. 

#### `plot_multi_rmse_curves()`: 

**FUNCTION** - plots the RMSE and minimum RMSE curves for MULTIPLE simulation. 

**ARGUMENTS** - basedir, filedir, sim_vect = simulations you want to plot, timestamp. Calls `get_reports()`. *WARNING: YOU WILL NEED TO MANUALLY EDIT THE FACTOR NAMES TO CORRESPOND TO THE SIMS/GENS YOU WANT!!*

**WARNING** - RMSE values greater than 12 are exclused from the plot.  
```{r}

plot_multi_rmse_curves <- function(basedir, filedir, total_sims, sims_vector){

  x_scale <- c(0, 100, 200, 300, 400, 500)
  y_scale <- c(0, 3, 6, 9, 12)
  mycolors <- c("black", "#CC79A7")

  top_rmse <- get_reports(basedir, filedir, total_sims) %>% 
              filter(sim %in% sims_vector) %>% 
              mutate(sim_label = paste0("Trial ", sim)) %>% 
              mutate(sim_label = factor(sim_label, levels = c("Trial 79","Trial 94", "Trial 10"))) %>% 
              filter(!is.na(gen)) %>% 
              filter(gen > 0) %>% 
              filter(error <= 12) %>% 
              ggplot(aes(x = gen)) +
              facet_wrap(~sim_label) +
              geom_line(aes(y = error, color = "RMSE")) +
              geom_line(aes(y = min_error, color = "Minimum RMSE")) +
              labs(colour="") +
              xlab("Generation") +
              ylab("RMSE") + 
              scale_y_continuous(breaks = y_scale) + 
              scale_x_continuous(breaks = x_scale) + 
              scale_color_manual(values = mycolors) + 
              cowplot::theme_minimal_grid(12) + 
              theme(axis.text.x = element_text(size=12, angle= -45, family="Arial"),
                    axis.text.y = element_text(size=12, family="Arial"),
                    axis.title = element_text(size=12, family="Arial"),
                    strip.text.x = element_text(size = 12,  family="Arial"),
                    legend.text = element_text(size = 12,  family="Arial"))
    
  return(top_rmse)
}

my_best_sims <- c(79, 94, 10)
my_basedir <- "/stor/work/Wilke/tingle/phix174/output/4152022/"
my_wt_filedir <- ""
my_total_sims = 100

plot_multi_rmse_curves(my_basedir, my_wt_filedir, my_total_sims, my_best_sims)

```



#### `best_wt_runs_data()`: 

**FUNCTION** - returns a dataframe. Contains 1) all of the wildtype simulation runs at timepoint 1200, with transcript abundances normalized expression of gene A 2) It has the experimentally-derived, normalized transcript abundances described in Logel 2020 (1:6:17...etc)

**ARGUMENTS** - basedir, best_wt_sims (the simulation numbers you want to plot), best_wt_gens (the generation number within each simulation that you want to plot). 

**EXAMPLE** - Uses the same `basedir` in the code block underneath **STEP 1**. Also the best_wt_sims and best_wt_gens refer to results from the 4/15/2022 runs used in my thesis. This code is commented out. 

```{r}
best_wt_runs_data <- function(best_wt_sims, best_wt_gens, basedir, filedir){
  
  i = 1
  wt_data <- data.frame()
  while(i <= length(best_wt_sims)){
    
    get_data <- read_tsv(paste0(basedir, filedir,"sim_", best_wt_sims[i], "_gen_", best_wt_gens[i], ".tsv" )) %>% 
                mutate(time = round(time)) %>% 
                filter(time == 1200) %>% 
                filter(!str_detect(species, '__|bound|ecoli|promoter|degraded')) %>% 
                mutate(normalized = transcript/transcript[1]) %>% 
                mutate(species = factor(species, c("gene_A", "gene_A*", "gene_B",
                                                 "gene_K", "gene_C", "gene_D",
                                                 "gene_E", "gene_J", "gene_F",
                                                 "gene_G", "gene_H"),
                                               c("A", "A*", "B", "K",
                                                 "C", "D", "E", "J",
                                                 "F", "G", "H"))) %>% 
                mutate(type = paste0("Trial ", best_wt_sims[i])) %>% 
                select(species, normalized, type) %>% 
                mutate(genome = "Wildtype ΦX174")
    
    wt_data <- rbind(wt_data, get_data)
    i = i + 1
  }
  
  wt_experimental_data <- data.frame(species = c("A", "A*", "B", "K","C", "D", "E", "J","F", "G", "H"),
                          normalized = c(1,1,6,6,6,17,17,17,11,5,1),
                          type = rep("Experimental Data")) %>% 
                          mutate(species = factor(species, c("A", "A*", "B", "K",
                                           "C", "D", "E", "J",
                                           "F", "G", "H"),
                                            c("A", "A*", "B", "K",
                                           "C", "D", "E", "J",
                                           "F", "G", "H"))) %>% 
                          mutate(genome = "Wildtype ΦX174")
  
  wt_data <- rbind(wt_data, wt_experimental_data)
  
  return(wt_data)
}

basedir <- "/stor/work/Wilke/tingle/phix174/output/4152022/"
filedir <- ""
best_wt_sims <- c(79, 94, 10)
best_wt_gens <- c(468, 202, 233)
my_wt_runs_data <- best_wt_runs_data(best_wt_sims, best_wt_gens, basedir, filedir)
```

#### `best_wt_runs_plot()`: 

**FUNCTION** - Uses a dataframe outputted from `best_wt_runs_data()` to plot the relative abundances of each wildtype gene. This specific figure was not used in my thesis, but a version of it was used to create another figure. 

**ARGUMENTS** - wt_data is the output from `best_wt_runs_data()`. *YOU WILL NEED TO MANUALLY EDIT THE FACTOR NAMES TO CORRESPOND TO THE SIMS/GENS YOU WANT!!*

**EXAMPLE** - Code is commented out. Calls dataframe created in the example for `best_wt_runs_dat()`. 

```{r}
best_wt_runs_plot <- function(wt_data){
  wt_fig <- wt_data %>% 
            # EDIT THE FACTOR NAMES BELOW TO MATCH THE SIMS/GENS YOU WANT
            mutate(type = factor(type, c("Experimental Data", "Trial 79", "Trial 94", "Trial 10"),
                                         c("Experimental Data", "Trial 79", "Trial 94", "Trial 10"))) %>% 
            ggplot(aes(x = species, y = normalized, color = type, shape = type)) +
            geom_point(size = 2.5) + 
            scale_y_continuous(limits = c(0, 19),
                              breaks = c(0,3,6,9,12,15,18)) +
            labs(subtitle = "Wildtype ΦX174") +
            ylab("Relative Transcript Abundances") +
            xlab("ΦX174 genes") + 
            scale_color_manual(values = c("#e37c71", "#808080", "#BEBEBE","black")) +
            cowplot::theme_minimal_hgrid(12) +
            theme(plot.subtitle = element_text(hjust = 0.5, size = 12,family = "Arial"),
                  axis.text.x = element_text(size=12, family="Arial"),
                  axis.text.y = element_text(size=12, family="Arial"),
                  axis.title = element_text(size=12, family="Arial"),
                  strip.text.x = element_text(size = 12,  family="Arial"),
                  legend.title = element_blank()) 
  
  return(wt_fig)
}

best_wt_runs_plot(my_wt_runs_data)
```


#### `best_dc_runs_data()`
**FUNCTION** - returns a dataframe. Contains 1) all of the decompressed simulation runs at timepoint 1200, with transcript abundances normalized expression of gene A 2) It has the experimentally-derived, normalized transcript abundances for the WILDTYPE described in Logel 2020 (1:6:17...etc)

**ARGUMENTS** - basedir, decomp_filedir (path to decompressed runs), best_dc_sims (the simulation numbers you want to plot), best_dc_gens (the generation number within each simulation that you want to plot). 

**EXAMPLE** - Uses the same `basedir` in the code block underneath **STEP 1**. Also the best_dc_sims and best_dc_gens refer to results from the 4/15/2022 runs used in my thesis. This code is commented out. 
```{r}
best_dc_runs_data <- function(best_dc_sims, best_dc_gens, basedir, decomp_filedir){
  
  i = 1
  dc_data <- data.frame()
  while(i <= length(best_dc_sims)){
    
    get_data <- read_tsv(paste0(basedir, decomp_filedir,"sim_", best_dc_sims[i], "_gen_", best_dc_gens[i], ".tsv" )) %>% 
                mutate(time = round(time)) %>% 
                filter(time == 1200) %>% 
                filter(!str_detect(species, '__|bound|ecoli|promoter|degraded')) %>% 
                mutate(normalized = transcript/transcript[1]) %>% 
                mutate(species = factor(species, c("gene_A", "gene_A*", "gene_B",
                                                 "gene_K", "gene_C", "gene_D",
                                                 "gene_E", "gene_J", "gene_F",
                                                 "gene_G", "gene_H"),
                                               c("A", "A*", "B", "K",
                                                 "C", "D", "E", "J",
                                                 "F", "G", "H"))) %>% 
                mutate(type = paste0("Trial ", best_dc_sims[i])) %>% 
                select(species, normalized, type) %>% 
                mutate(genome = "Decompressed ΦX174")
    
    dc_data <- rbind(dc_data, get_data)
    i = i + 1
  }
  
  wt_experimental_data <- data.frame(species = c("A", "A*", "B", "K","C", "D", "E", "J","F", "G", "H"),
                          normalized = c(1,1,6,6,6,17,17,17,11,5,1),
                          type = rep("Experimental Data")) %>% 
                          mutate(species = factor(species, c("A", "A*", "B", "K",
                                           "C", "D", "E", "J",
                                           "F", "G", "H"),
                                            c("A", "A*", "B", "K",
                                           "C", "D", "E", "J",
                                           "F", "G", "H"))) %>% 
                          mutate(genome = "Decompressed ΦX174")
  
  dc_data <- rbind(dc_data, wt_experimental_data)
  
  return(dc_data)
}

basedir <- "/stor/work/Wilke/tingle/phix174/output/4152022/"
filedir <- "decompressed_runs/"
best_dc_sims <- c(79, 94, 10)
best_dc_gens <- c(468, 202, 233)
my_dc_runs_data <- best_dc_runs_data(best_dc_sims, best_dc_gens, basedir, filedir)
```

#### `best_dc_runs_plot()`: 

**FUNCTION** - Uses a dataframe outputted from `best_dc_runs_data()` to plot the relative abundances of each decompressed gene. This specific figure was not used in my thesis, but a version of it was used to create another figure. 

**ARGUMENTS** - dc_data is the output from `best_dc_runs_data()`. *YOU WILL NEED TO MANUALLY EDIT THE FACTOR NAMES TO CORRESPOND TO THE SIMS/GENS YOU WANT!!* ALSO! The scale_y_continuous() values used to create the figure may need to be manually changed. 

**EXAMPLE** - Code is commented out. Calls dataframe created in the example for `best_dc_runs_dat()`. 

```{r}
best_dc_runs_plot <- function(dc_data){
  dc_fig <- dc_data %>% 
            # EDIT THE FACTOR NAMES BELOW TO MATCH THE SIMS/GENS YOU WANT
            mutate(type = factor(type, c("Experimental Data", "Trial 79", "Trial 94", "Trial 10"),
                                         c("Experimental Data", "Trial 79", "Trial 94", "Trial 10"))) %>% 
            ggplot(aes(x = species, y = normalized, color = type, shape = type)) +
            geom_point(size = 2.5) + 
            scale_y_continuous(limits = c(0, 36),
                                breaks = c(0,3,6,9,12,15,18,21,24,27,30,33)) +
            labs(subtitle = "Decompressed ΦX174") +
            ylab("Relative Transcript Abundances") +
            xlab("ΦX174 genes") + 
            scale_color_manual(values = c("#e37c71", "#808080", "#BEBEBE","black")) +
            cowplot::theme_minimal_hgrid(12) +
            theme(plot.subtitle = element_text(hjust = 0.5, size = 12,family = "Arial"),
                  axis.text.x = element_text(size=12, family="Arial"),
                  axis.text.y = element_text(size=12, family="Arial"),
                  axis.title = element_text(size=12, family="Arial"),
                  strip.text.x = element_text(size = 12,  family="Arial"),
                  legend.title = element_blank()) 
  
  return(dc_fig)
}

best_dc_runs_plot(my_dc_runs_data)
```


#### `compare_wt_dc_transcripts_plot()` 
**FUNCTION** - Uses dataframes outputted from `best_wt_runs_data()` and `best_dc_runs_data()` to plot the relative abundances of each wildtype and decompressed gene. This is Figure 5 in Tanvi's UG thesis. 

**ARGUMENTS** - wt_data is the output from `best_wt_runs_data()`. dc_data is the output from `best_dc_runs_data()`. *YOU WILL NEED TO MANUALLY EDIT THE FACTOR NAMES TO CORRESPOND TO THE SIMS/GENS YOU WANT!!* ALSO! The scale_y_continuous() values used to create the figure may need to be manually changed. 

**EXAMPLE** - Code is commented out. Calls dataframe created in the example for `best_wt_runs_data()` and `best_dc_runs_data()`. 

```{r}
compare_wt_dc_transcripts_plot <- function(wt_data, dc_data) {
  
  all_data <- rbind(wt_data, dc_data)

  all_figure <- all_data %>% 
                mutate(type = factor(type, c("Experimental Data", "Trial 79", "Trial 94", "Trial 10"),
                                                         c("Experimental Data", "Trial 79", "Trial 94", "Trial 10"))) %>% 
                mutate(genome = factor(genome, c("Wildtype ΦX174", "Decompressed ΦX174"),
                                                         c("Wildtype ΦX174", "Decompressed ΦX174"))) %>% 
                ggplot(aes(x = species, y = normalized, color = type, shape = type)) +
                facet_wrap(~genome) +
                geom_point(size = 2.5) + 
                scale_y_continuous(limits = c(0, 36),
                                   breaks = c(0,3,6,9,12,15,18,21,24,27,30,33)) +
                ylab("Relative Transcript Abundances") +
                xlab("ΦX174 genes") + 
                scale_color_manual(values = c("#e37c71","#808080", "#BEBEBE","black")) +
                cowplot::theme_minimal_hgrid(12) +
                theme(plot.subtitle = element_text(hjust = 0.5, size = 12,family = "Arial"),
                      axis.text.x = element_text(size=12, family="Arial"),
                      axis.text.y = element_text(size=12, family="Arial"),
                      axis.title = element_text(size=12, family="Arial"),
                      strip.text.x = element_text(size = 12,  family="Arial"),
                      legend.title = element_blank()) 
  
  return(all_figure)
                
}

compare_wt_dc_transcripts_plot(my_wt_runs_data, my_dc_runs_data)
```

#### `compare_wt_dc_absolute_transcripts_plot()`
**FUNCTION** - Plots absolute transcript abundances for each gene, faceted by wildtype or decompressed genomes. 

**ARGUMENTS** - best_sims = vector with best simulation numbers, best_gens = vector with best generations within each simulation, basedir, wt_filedir = directory with wildtype runs, dc_filedir = directory with decompressed runs. *YOU WILL NEED TO MANUALLY EDIT THE FACTOR NAMES TO CORRESPOND TO THE SIMS/GENS YOU WANT!!* 

```{r}
compare_wt_dc_absolute_transcripts_plot <- function(best_sims, best_gens, basedir, wt_filedir, dc_filedir) {
  
  # Get wt data w/ absolute counts for specific simulations
  i = 1
  wt_data <- data.frame()
  while(i <= length(best_sims)){
    
  get_data <- read_tsv(paste0(basedir, wt_filedir,"sim_", 
                              best_sims[i], "_gen_", best_gens[i], ".tsv" )) %>% 
                mutate(time = round(time)) %>% 
                filter(time == 1200) %>% 
                filter(!str_detect(species, '__|bound|ecoli|promoter|degraded')) %>% 
                mutate(species = factor(species, c("gene_A", "gene_A*", "gene_B",
                                                 "gene_K", "gene_C", "gene_D",
                                                 "gene_E", "gene_J", "gene_F",
                                                 "gene_G", "gene_H"),
                                               c("A", "A*", "B", "K",
                                                 "C", "D", "E", "J",
                                                 "F", "G", "H"))) %>% 
                mutate(type = paste0("Trial ", best_sims[i])) %>% 
                select(species, transcript, type) %>% 
                mutate(genome = "Wildtype ΦX174")
    
    wt_data <- rbind(wt_data, get_data)
    i = i + 1
  }
  
  # Get dc data w/ absolute counts for specific simulations
  j = 1
  dc_data <- data.frame()
  while(j <= length(best_sims)){
    
    get_data <- read_tsv(paste0(basedir, dc_filedir,"sim_", 
                                best_sims[j], "_gen_", best_gens[j], ".tsv" )) %>% 
                mutate(time = round(time)) %>% 
                filter(time == 1200) %>% 
                filter(!str_detect(species, '__|bound|ecoli|promoter|degraded')) %>% 
                mutate(species = factor(species, c("gene_A", "gene_A*", "gene_B",
                                                 "gene_K", "gene_C", "gene_D",
                                                 "gene_E", "gene_J", "gene_F",
                                                 "gene_G", "gene_H"),
                                               c("A", "A*", "B", "K",
                                                 "C", "D", "E", "J",
                                                 "F", "G", "H"))) %>% 
                mutate(type = paste0("Trial ", best_sims[j])) %>% 
                select(species, transcript, type) %>% 
                mutate(genome = "Decompressed ΦX174")
    
    dc_data <- rbind(dc_data, get_data)
    j = j + 1
  }
  
  # Combine datasets into 1 big dataset 
  all_abs_data <- rbind(wt_data, dc_data)
  
  # Plot 
  all_figure <- all_abs_data %>% 
                mutate(type = factor(type, c("Trial 79", "Trial 94", "Trial 10"),
                                           c("Trial 79", "Trial 94", "Trial 10"))) %>% 
                mutate(genome = factor(genome, c("Wildtype ΦX174", "Decompressed ΦX174"),
                                               c("Wildtype ΦX174", "Decompressed ΦX174"))) %>% 
                ggplot(aes(x = species, y = transcript, color = type, shape = type)) +
                facet_wrap(~genome) +
                geom_point(size = 2.5) + 
                scale_shape_manual(values = c(17, 15, 3)) + 
                #geom_point(size = 2.5, shape = c(rep(c(17), 11),rep(c(15), 11),rep(c(3), 11))) + 
                #scale_y_continuous(limits = c(0, 36),
                #                   breaks = c(0,3,6,9,12,15,18,21,24,27,30,33)) +
                ylab("Absolute Transcript Abundances") +
                xlab("ΦX174 genes") + 
                scale_color_manual(values = c("#808080", "#BEBEBE","black")) +
                cowplot::theme_minimal_hgrid(12) +
                theme(plot.subtitle = element_text(hjust = 0.5, size = 12,family = "Arial"),
                      axis.text.x = element_text(size=12, family="Arial"),
                      axis.text.y = element_text(size=12, family="Arial"),
                      axis.title = element_text(size=12, family="Arial"),
                      strip.text.x = element_text(size = 12,  family="Arial"),
                      legend.title = element_blank()) 
  
  return(all_figure)
}

my_best_sims <- c(79, 94, 10)
my_best_gens <- c(468, 202, 233)
my_basedir <- "/stor/work/Wilke/tingle/phix174/output/4152022/"
my_wt_filedir <- ""
my_dc_filedir <- "decompressed_runs/"
compare_wt_dc_absolute_transcripts_plot(my_best_sims, my_best_gens, my_basedir, my_wt_filedir, my_dc_filedir) 
```


#### `plot_fitness()`
**FUNCTION** - creates Figure 7 from Tanvi's UG thesis. Plots barplot with experimental v simulated number of total virions (as calculated by the stoichiometric relationship between translation levels of structural proteins). 

**ARGUMENTS** - best_sims = vector with best simulation numbers, best_gens = vector with best generations within each simulation, basedir, wt_filedir = directory with wildtype runs, dc_filedir = directory with decompressed runs. *YOU WILL NEED TO MANUALLY EDIT THE FACTOR NAMES & NUMBER YOU WANT TO CORRESPOND TO THE SIMS YOU WANT!!* 

**WARNING** - `plot_fitness()` gives EITHER the wildtype or the decompressed panels from Figure 7 in Tanvi's UG thesis. See code below function definition for how to combine the two plots into 1 figure. 

```{r}
plot_fitness <- function(basedir, filedir, best_sims, best_gens, timestamp, genome, subtitle){
  i = 1
  all_virions <- data.frame()
  
  if(genome == "wildtype"){
    all_virions <- data.frame(virions = c(145), # experimentally-derived, published number of new virions
                            data_type = c("Experimental"),
                            trial = c("Experimental"),
                            ymin = c(82),
                            ymax = c(208)) 
  }
  
  if(genome == "decompressed"){
    all_virions <- data.frame(virions = c(16), # experimentally-derived, published number of new virions
                            data_type = c("Experimental"),
                            trial = c("Experimental"),
                            ymin = c(14),
                            ymax = c(18)) 
  }
  
               
  while (i <= length(best_sims)){
    
    new_virions <- get_fitness(basedir, filedir, best_sims[i], best_gens[i], timestamp)
    new_row <- data.frame(virions = new_virions,
                          data_type = "Simulation",
                          trial = paste0("Trial ", best_sims[i]),
                          ymin = 5,
                          ymax = 5)
    
    all_virions <- rbind(all_virions, new_row)
    i = i + 1
  }
  
  i = 1
  all_virions <- all_virions %>% 
                 mutate(trial = factor(trial, 
                                       c("Experimental", paste0("Trial ", best_sims[i]), 
                                                         paste0("Trial ", best_sims[i + 1]),
                                                         paste0("Trial ", best_sims[i + 2])),
                                       c("Experimental", paste0("Trial ", best_sims[i]), 
                                                         paste0("Trial ", best_sims[i + 1]),
                                                         paste0("Trial ", best_sims[i + 2]))))
  
  fitness_plot <- all_virions %>% 
                  ggplot(aes(x = trial, y = virions, fill = data_type)) +
                  geom_bar(stat="identity", position=position_dodge()) +
                  geom_errorbar(aes(ymin = ymin, ymax=ymax), 
                              color = c("black", "#E69F00", "#E69F00", "#E69F00"),
                              width = 0.2, position = position_dodge(0.8)) +
                  xlab("ΦX174 model") +
                  ylab("ΦX174 virions after lysis") + 
                  labs(subtitle = subtitle)  + 
                  scale_y_continuous(limits = c(0, 210),
                                   breaks = c(0,50, 100, 150, 200)) +
                  scale_fill_manual(values=c("#50C4DE", "#E69F00")) +
                  cowplot::theme_cowplot(12) +
                  theme(legend.position = "none",
                      plot.subtitle = element_text(hjust = 0.5, size = 12,family = "Arial"),
                      axis.text.x = element_text(size=12, family="Arial"),
                      axis.text.y = element_text(size=12, family="Arial"),
                      axis.title = element_text(size=12, family="Arial"),
                      strip.text.x = element_text(size = 12,  family="Arial"))
  
  return(fitness_plot)
}

my_best_sims <- c(79, 94, 10)
my_best_gens <- c(468, 202, 233)
my_basedir <- "/stor/work/Wilke/tingle/phix174/output/4152022/"
my_wt_filedir <- ""
my_dc_filedir <- "decompressed_runs/"
timestamp = 1200

wt_fitness <- plot_fitness(my_basedir, my_wt_filedir, my_best_sims, my_best_gens, timestamp, "wildtype", "Wildtype ΦX174 Fitness")

dc_fitness <- plot_fitness(my_basedir, my_dc_filedir, my_best_sims, my_best_gens, timestamp, "decompressed" ,"Decompressed ΦX174 Fitness")

plot_grid(wt_fitness, dc_fitness, ncol=2)
```
