---
title: "2102022_analysis"
author: "Tanvi Ingle" 
date: "2/11/2022"
output: html_document
---
Libraries 
```{r, echo = FALSE, message = FALSE, warning=FALSE}
require(ggplot2)
require(tidyverse)
require(cowplot)
require(scales)
require(kableExtra)
```

Common Functions & Base directory 
```{r, echo = FALSE, message = FALSE, warning=FALSE}
# Functions & Variables I use a lot
basedir <- "/stor/work/Wilke/tingle/phix174/output/2102022/"

# get_reports() retrieves the reports for all the simulation runs in a specified scenario
# had hardcoded case for Scenario B where there is no simulation 1
get_reports <- function(basedir, filedir, total_sims){
  report <- data.frame()
  #if(filedir == "scenario_B/") {
  #  i = 2
  #}
  #else 
  i = 1
  while (i<=total_sims) {
    sim = read_csv(paste0(basedir, filedir, "sim_", i, "_report.csv")) 
    sim$sim = i
    report <- rbind(report, sim)
    i <- i + 1
  }
  report <- report %>% 
    select(-X1)
  return(report)
}

# plot_reports() plots the RMSE v generations for all the simulations in a given scenario
plot_reports <- function(basedir, filedir, total_sims, title){
  figure <- get_reports(basedir, filedir, total_sims) %>% 
    filter(!is.na(gen)) %>% 
    filter(gen > 0) %>% 
    filter(error < 11) %>% 
    ggplot(aes(x = gen, y = error)) +
    geom_line() + 
    facet_wrap(~sim) + 
    ggtitle(title) + 
    xlab("Generations") +
    ylab("RMSE") + 
    ylim(c(1, 10)) + 
    cowplot::theme_minimal_grid(12)
  
  return(figure)
}

# logel_plot() recreates the barplot transcript abundances from Logel 2020
# sim = simulation number, gen = generation of interest within the simulation, timestamp = pinetree run
logel_plot <- function(basedir, filedir, sim, gen, timestamp){
  
  my_palette = c("#d0a34f","#75c130", "#c12c04", "#ecd575", "#b69507",
            "#e37c71", "#adcb40", "#d3ad40", "#288ede", "#a5bdfa",
            "#93c9ff")

  figure <- read_tsv(paste0(basedir, filedir,"sim_", sim, "_gen_", gen, "_ptrun.tsv" )) %>% 
       mutate(time = round(time)) %>% 
       filter(time == timestamp) %>% 
       filter(!str_detect(species, '__|bound|ecoli|promoter')) %>% 
       mutate(normalized = transcript/transcript[1]) %>% 
       mutate(species = factor(species, c("gene_A", "gene_A*", "gene_B",
                                         "gene_K", "gene_C", "gene_D",
                                         "gene_E", "gene_J", "gene_F",
                                         "gene_G", "gene_H"),
                                       c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"))) %>% 
       ggplot(aes(x=species, y=normalized, fill = species)) +
       geom_bar(stat="identity", color = "black") + 
       scale_fill_manual(values = my_palette) + 
       scale_y_continuous(limits = c(0, 10), 
                          breaks = c(0,1, 2, 3, 4, 5, 6 ,7, 8, 9, 10)) + 
       #ylab("Transcript abundances") + 
       ylab("")+ 
       # geom_hline(yintercept=1, linetype="dashed", color = "red") + 
       # geom_hline(yintercept=4.5, linetype="dashed", color = "red") + 
       # geom_hline(yintercept=4.7, linetype="dashed", color = "red") + 
       # geom_hline(yintercept=8.5, linetype="dashed", color = "red") + 
       # geom_hline(yintercept=7.5, linetype="dashed", color = "red") + 
       # geom_hline(yintercept=8.7, linetype="dashed", color = "red") + 
       # geom_hline(yintercept=4.6, linetype="dashed", color = "red") + 
       # geom_hline(yintercept=2.9, linetype="dashed", color = "red") + 
       # geom_hline(yintercept=9.8, linetype="dashed", color = "red") + 
       # geom_hline(yintercept=3, linetype="dashed", color = "red") + 
       cowplot::theme_cowplot(12) +
       theme(legend.position = "none") 
  
  return(figure)
}

# logeL_timeseries called logel_plot() to show how distribution changes over time (start, middle, end)
# sim = simulation number, gen = last generation of interest, timestamp = pinetree run

logel_timeseries <- function(basedir, filedir, sim, gen, timestamp) {
  
  sim_start <- logel_plot(basedir, filedir, sim, gen = 1, timestamp = 1200)+
    ylab(paste0("Sim ", sim))
  sim_middle <- logel_plot(basedir, filedir, sim, gen = gen/2, timestamp = 1200)
  sim_end <- logel_plot(basedir, filedir, sim, gen = gen, timestamp = 1200)
  
  figure <- plot_grid(sim_start, sim_middle, sim_end, 
                      label_x = "genes", label_y = "transcript abundances", 
                      ncol = 3, nrow = 1)
                     
  return(figure)
}
```

### **Objective**
Here I will analyze the results from the promoter optimization run on 2-10-2022. For this optimization attempt, I ran 5 scenarios with unique distributions for promoters A, B, and D. Each scenario was simmulated 10 times. Each simulation was run for 300 generations. 

The optimization algorithm differs from that run on 2-08-2022 in one main way: 

1. The data are being fit to RNAseq data instead of qPCR data. Since this is a more complicated distribution, I am not sure if my current optimization algorithm can fit to this. 

Here are the details of each run. Each value represents the exponent on *e*, ie if uA = 11, then the mean strength of promoter A is e^11.

> Starting promoter distributions

```{r, echo = FALSE, message = FALSE, warning=FALSE}
scenarios <- data.frame(scenario = c("A", "B", "C", "D", "E"), 
                        uA = c(11, 11.8, 12.21, 12.8, 13.2),
                        oA = c(10, 10, 10, 10, 10),
                        uB = c(12.8, 13.5, 14.33, 14.8, 15.4), 
                        oB = c(10, 12, 12, 12, 12),
                        uD = c(12.3, 12.8, 14.11, 14.5, 14.8),
                        oD = c(11, 11, 11, 11, 11))
kable(scenarios) %>% 
   kable_styling(latex_options = "striped")
```

### **Scenario A**

> Scenario A - Change in Error over Generations

```{r, echo = FALSE, message = FALSE, warning=FALSE}
filedir <- "scenario_A/"
scenario_A <- plot_reports(basedir, filedir, 10, "RNAseq: Scenario A")
scenario_A
```

> Scenario A - Associated Errors for the best simulations

```{r, echo = FALSE, message = FALSE, warning=FALSE}
col_order = c("sim", "gen", "pA", "pB", "pD", "error")

scenario_A_report <- get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(3,8)) %>% 
  group_by(sim) %>% 
  mutate(min_error = min(error)) %>% 
  filter(gen == 300 | error == min_error) %>% 
  filter(row_number()==1 | gen == 300)

scenario_A_report <- scenario_A_report[ , col_order]

kable(scenario_A_report) %>% 
  kable_styling(latex_options = "striped") %>% 
  row_spec(1, bold = T )
   
```

> Scenario A - Final Transcript Abundances for the best sims 

```{r, echo = FALSE, message = FALSE, warning=FALSE}
logel_report <- scenario_A_report %>% 
  select(sim, gen) 

sim3_280 <- logel_plot(basedir, filedir, 3, 280, 1200) + 
  ggtitle("sim 3, gen = 280")

sim3_300 <- logel_plot(basedir, filedir, 3, 300, 1200) + 
  ggtitle("sim 3, gen = 300")

sim8_102 <- logel_plot(basedir, filedir, 8, 102, 1200) +
  theme(axis.title.y = element_blank()) +
  ggtitle("sim 8, gen = 102")
  
sim8_300 <- logel_plot(basedir, filedir, 8, 300, 1200) +
  theme(axis.title.y = element_blank()) +
  ggtitle("sim 8, gen = 300")

scenario_A_best <- plot_grid(sim3_280, sim3_300, 
                             sim8_102, sim8_300,
                             ncol = 2, nrow = 2)

scenario_A_best
```

> Scenario A - Evolution of transcript abundances over generations for the best sims

```{r, echo = FALSE, message = FALSE, warning=FALSE}
sim5_time <- logel_timeseries(basedir, filedir, 5, 300, 1200) 
sim6_time <- logel_timeseries(basedir, filedir, 6, 300, 1200) 
sim8_time <- logel_timeseries(basedir, filedir, 8, 300, 1200) 
#sim10_time <- logel_timeseries(basedir, filedir, 10, 300, 1200)

plot_grid(sim5_time, sim6_time, sim8_time, nrow = 3)
```

> Scenario A - Here is how the promoter values change over time 

```{r, echo = FALSE, message = FALSE, warning=FALSE}
get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(5,6,8)) %>% 
  #filter(gen <= 201) %>% 
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim) +
  geom_line(aes(y = pA)) + 
  scale_y_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  cowplot::theme_minimal_grid(12)

get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(5,6,8)) %>%
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim) +
  scale_y_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  geom_line(aes(y = pB)) + 
  cowplot::theme_minimal_grid(12)


get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(5,6,8)) %>%
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim, scales="free_y") +
  geom_line(aes(y = pD)) + 
  scale_y_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  cowplot::theme_minimal_grid(12)
  
  # geom_line(aes(y = pB)) + 
  # geom_line(aes(y = pD))
```

-----------------------------------------------------------------------------------
### **Scenario B**

> Scenario B - Change in Error over Generations

```{r, echo = FALSE, message = FALSE, warning=FALSE}
filedir <- "scenario_B/"
scenario_B <- plot_reports(basedir, filedir, 10, "RNAseq: Scenario B")
scenario_B
```

> Scenario B - Associated Errors for the best simulations

```{r, echo = FALSE, message = FALSE, warning=FALSE}
col_order = c("sim", "gen", "pA", "pB", "pD", "error")

scenario_B_report <- get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(2, 4, 8)) %>% 
  group_by(sim) %>% 
  mutate(min_error = min(error)) %>% 
  filter(gen == 300 | error == min_error) %>% 
  filter(row_number()== 1 | gen == 300)

scenario_B_report <- scenario_B_report[ , col_order] 

kable(scenario_B_report) %>% 
  kable_styling(latex_options = "striped") %>% 
  row_spec(3, bold = T )
```

> Scenario B - Final Transcript Abundances for the best sims 

```{r, echo = FALSE, message = FALSE, warning=FALSE}
sim2_84 <- logel_plot(basedir, filedir, 2, 84, 1200) + 
  ggtitle("sim 2, gen = 84") 

# sim2_300 <- logel_plot(basedir, filedir, 2, 300, 1200) + 
#   ggtitle("sim 2, gen = 300") 

sim4_27 <- logel_plot(basedir, filedir, 4, 27, 1200) + 
  ggtitle("sim 4, gen = 27") 

sim8_92 <- logel_plot(basedir, filedir, 8, 92, 1200) + 
  ggtitle("sim 8, gen = 92") 

scenario_B_best <- plot_grid(sim2_84, sim4_27, sim8_92,ncol = 3)

scenario_B_best
```

> Scenario B - Evolution of transcript abundances over generations for the best sims

```{r, echo = FALSE, message = FALSE, warning=FALSE}
sim9_time <- logel_timeseries(basedir, filedir, 9, 300, 1200) 
#sim8_time <- logel_timeseries(basedir, filedir, 8, 300, 1200) 

sim8_start <- logel_plot(basedir, filedir, 8, gen = 1, timestamp = 1200)
sim8_middle <- logel_plot(basedir, filedir, 8, gen = 150, timestamp = 1195)
sim8_end <- logel_plot(basedir, filedir, 8, gen = 300, timestamp = 1200)
  
sim8_time <- plot_grid(sim8_start, sim8_middle, sim8_end, 
                      label_x = "genes", label_y = "transcript abundances",
                      ncol = 3, nrow = 1)
                     
#sim10_time <- logel_timeseries(basedir, filedir, 10, 300, 1200)

plot_grid(sim9_time, sim8_time, nrow = 2)
```

> Scenario B - Here is how the promoter values change over time 

```{r, echo = FALSE, message = FALSE, warning=FALSE}
get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(8,9)) %>% 
  #filter(gen <= 201) %>% 
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim) +
  geom_line(aes(y = pA)) + 
  scale_y_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  cowplot::theme_minimal_grid(12)

get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(8,9)) %>%
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim) +
  scale_y_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  geom_line(aes(y = pB)) + 
  cowplot::theme_minimal_grid(12)


get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(8,9)) %>%
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim) +
  geom_line(aes(y = pD)) + 
  scale_y_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  cowplot::theme_minimal_grid(12)
```

-----------------------------------------------------------------------------------
### **Scenario C**

> Scenario C - Change in Error over Generations

```{r, echo = FALSE, message = FALSE, warning=FALSE}
filedir <- "scenario_C/"
scenario_C <- plot_reports(basedir, filedir, 10, "RNAseq: Scenario C")
scenario_C
```

> Scenario C - Associated Errors for the best simulations

```{r, echo = FALSE, message = FALSE, warning=FALSE}
col_order = c("sim", "gen", "pA", "pB", "pD", "error")

scenario_C_report <- get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(2,3,8,10)) %>% 
  group_by(sim) %>% 
  mutate(min_error = min(error)) %>% 
  filter(gen == 300 | error == min_error) %>% 
  filter(row_number()== 1 | gen == 300) %>% 
  ungroup() %>% 
  select(-min_error) %>% 
  arrange(error)

scenario_C_report <- scenario_C_report[ , col_order] 

kable(scenario_C_report) %>% 
  kable_styling(latex_options = "striped") %>% 
  row_spec(1:3, bold = T )

```

> Scenario C - Final Transcript Abundances for the best sims 

```{r, echo = FALSE, message = FALSE, warning=FALSE}
# my_palette = c("#d0a34f","#75c130", "#c12c04", "#ecd575", "#b69507",
#             "#e37c71", "#adcb40", "#d3ad40", "#288ede", "#a5bdfa",
#             "#93c9ff")
# logel_RNAseq <- data.frame(species = c("gene_A", "gene_A*", "gene_B", 
#                                        "gene_K","gene_C", "gene_D", 
#                                        "gene_E", "gene_J","gene_F", 
#                                        "gene_G", "gene_H"),
#                            abundances = c(1, 1, 4.5, 4.7, 8.5, 7.5, 8.7, 4.6,  2.9, 9.8, 3)) %>% 
#   mutate(species = factor(species, c("gene_A", "gene_A*", "gene_B",
#                                          "gene_K", "gene_C", "gene_D",
#                                          "gene_E", "gene_J", "gene_F",
#                                          "gene_G", "gene_H"),
#                                        c("A", "A*", "B", "K",
#                                          "C", "D", "E", "J",
#                                          "F", "G", "H"))) %>% 
#   ggplot(aes(x = species, y = abundances, fill = species)) + 
#   geom_bar(stat="identity", color = "black") + 
#        scale_fill_manual(values = my_palette) + 
#        scale_y_continuous(limits = c(0, 10), 
#                           breaks = c(0,1, 2, 3, 4, 5, 6 ,7, 8, 9, 10)) + 
#        ylab("")+ 
#        cowplot::theme_cowplot(12) +
#        theme(legend.position = "none") 
#   

sim3_16 <- logel_plot(basedir, filedir, 3, 16, 1200) + 
  ggtitle("sim 3, gen = 16, RMSE = 1.00875") 

sim10_125 <- logel_plot(basedir, filedir, 10, 125, 1200) + 
  ggtitle("sim 10, gen = 125, RMSE = 1.31320") 

sim3_300 <- logel_plot(basedir, filedir, 3, 300, 1200) + 
  ggtitle("sim 3, gen = 300, RMSE = 1.80754") 

sim8_300 <- logel_plot(basedir, filedir, 8, 300, 1200) + 
  ggtitle("sim 8, gen = 300, RMSE = 4.52408") 



scenario_C_best <- plot_grid(sim3_16, sim10_125, sim3_300, sim8_300, ncol= 2, nrow = 2)
scenario_C_best
```

> Scenario C - Evolution of transcript abundances over generations for the best sims

```{r, echo = FALSE, message = FALSE, warning=FALSE}
sim3_time <- logel_timeseries(basedir, filedir, 3, 300, 1200) 
sim8_time <- logel_timeseries(basedir, filedir, 8, 300, 1200) 
sim10_time <- logel_timeseries(basedir, filedir, 10, 300, 1200) 

plot_grid(sim3_time, sim8_time, sim10_time, nrow = 3)
```

> Scenario C - Here is how the promoter values change over time 

```{r, echo = FALSE, message = FALSE, warning=FALSE}
get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(8,10,2)) %>% 
  #filter(gen <= 201) %>% 
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim) +
  geom_line(aes(y = pA)) + 
  scale_y_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  cowplot::theme_minimal_grid(12)

get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(8,10,2)) %>%
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim) +
  scale_y_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  geom_line(aes(y = pB)) + 
  cowplot::theme_minimal_grid(12)


get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(8,10,2)) %>%
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim) +
  geom_line(aes(y = pD)) + 
  scale_y_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  cowplot::theme_minimal_grid(12)
```

-----------------------------------------------------------------------------------
### **Scenario D**

> Scenario D - Change in Error over Generations

```{r, echo = FALSE, message = FALSE, warning=FALSE}
filedir <- "scenario_D/"
scenario_D <- plot_reports(basedir, filedir, 10, "RNAseq: Scenario D")
scenario_D
```

> Scenario D - Associated Errors for the best simulations

```{r, echo = FALSE, message = FALSE, warning=FALSE}
col_order = c("sim", "gen", "pA", "pB", "pD", "error")

scenario_D_report <- get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(3,5)) %>% 
  group_by(sim) %>% 
  mutate(min_error = min(error)) %>% 
  filter(gen == 300 | error == min_error) %>% 
  filter(row_number()== 1 | gen == 300)

scenario_D_report <- scenario_D_report[ , col_order] 

kable(scenario_D_report) %>% 
  kable_styling(latex_options = "striped") %>% 
  row_spec(3, bold = T )
```

> Scenario D - Final Transcript Abundances for the best sims 

```{r, echo = FALSE, message = FALSE, warning=FALSE}
sim3_112 <- logel_plot(basedir, filedir, 3, 112, 1200) + 
  ggtitle("sim 3, gen = 112") 

sim3_300 <- logel_plot(basedir, filedir, 3, 300, 1200) + 
  ggtitle("sim 3, gen = 300") 

sim5_0 <- logel_plot(basedir, filedir, 5, 0, 1200) + 
  ggtitle("sim 5, gen = 0") 

sim5_300 <- logel_plot(basedir, filedir, 5, 300, 1200) + 
  ggtitle("sim 5, gen = 300") 

scenario_D_best <- plot_grid(sim3_112, sim3_300, sim5_0, sim5_300, ncol= 2, nrow = 2)
scenario_D_best
```

> Scenario D - Evolution of transcript abundances over generations for the best sims

```{r, echo = FALSE, message = FALSE, warning=FALSE}
sim3_time <- logel_timeseries(basedir, filedir, 3, 300, 1200) 
sim5_time <- logel_timeseries(basedir, filedir, 5, 300, 1200) 

plot_grid(sim3_time, sim5_time, nrow = 2)
```

> Scenario D - Here is how the promoter values change over time 

```{r, echo = FALSE, message = FALSE, warning=FALSE}
get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(3,5)) %>% 
  #filter(gen <= 201) %>% 
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim) +
  geom_line(aes(y = pA)) + 
  scale_y_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  cowplot::theme_minimal_grid(12)

get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(3,5)) %>%
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim) +
  scale_y_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  geom_line(aes(y = pB)) + 
  cowplot::theme_minimal_grid(12)


get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(3,5)) %>%
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim) +
  geom_line(aes(y = pD)) + 
  scale_y_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  cowplot::theme_minimal_grid(12)
```

-----------------------------------------------------------------------------------
### **Scenario E**

> Scenario E - Change in Error over Generations

```{r, echo = FALSE, message = FALSE, warning=FALSE}
filedir <- "scenario_E/"
scenario_E <- plot_reports(basedir, filedir, 10, "Figure 5: Scenario E")
scenario_E
```

> Scenario E - Associated Errors for the best simulations

```{r, echo = FALSE, message = FALSE, warning=FALSE}
col_order = c("sim", "gen", "pA", "pB", "pD", "error")

scenario_E_report <- get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(1,2,10)) %>% 
  group_by(sim) %>% 
  mutate(min_error = min(error)) %>% 
  filter(gen == 300 | error == min_error) %>% 
  filter(row_number()== 1 | gen == 300)

scenario_E_report <- scenario_E_report[ , col_order] 

kable(scenario_E_report) %>% 
  kable_styling(latex_options = "striped") %>% 
  row_spec(1, bold = T )
```

> Scenario E - Final Transcript Abundances for the best sims 

```{r, echo = FALSE, message = FALSE, warning=FALSE}
sim1_142 <- logel_plot(basedir, filedir, 1, 142, 1200) + 
  ggtitle("sim 1, gen = 142") 

sim2_270 <- logel_plot(basedir, filedir, 2, 270, 1200) + 
  ggtitle("sim 2, gen = 270") 

sim2_300 <- logel_plot(basedir, filedir, 2, 300, 1200) + 
  ggtitle("sim 2, gen = 300") 


scenario_E_best <- plot_grid(sim10_6, sim2_270, sim2_300, ncol= 3)
scenario_E_best
```

> Scenario E - Evolution of transcript abundances over generations for the best sims

```{r, echo = FALSE, message = FALSE, warning=FALSE}
sim10_time <- logel_timeseries(basedir, filedir, 10, 300, 1200) 
sim2_time <- logel_timeseries(basedir, filedir, 2, 300, 1200) 

plot_grid(sim10_time, sim2_time, nrow = 2)
```

> Scenario E - Here is how the promoter values change over time 

```{r, echo = FALSE, message = FALSE, warning=FALSE}
get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(2,10)) %>% 
  #filter(gen <= 201) %>% 
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim) +
  geom_line(aes(y = pA)) + 
  scale_y_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  cowplot::theme_minimal_grid(12)

get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(2,10)) %>%
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim) +
  scale_y_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  geom_line(aes(y = pB)) + 
  cowplot::theme_minimal_grid(12)

get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(2,10)) %>%
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim) +
  geom_line(aes(y = pD)) + 
  scale_y_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  cowplot::theme_minimal_grid(12)
```


