---
title: "3212022_analysis"
author: "Tanvi Ingle"
date: "3/22/2022"
output: html_document
---

Libraries 
```{r, echo = FALSE, message = FALSE, warning=FALSE}
require(ggplot2)
require(tidyverse)
require(cowplot)
require(scales)
require(kableExtra)
```

Common Functions & Base directory 
```{r, echo = FALSE, message = FALSE, warning=FALSE}
# Functions & Variables I use a lot
basedir <- "/stor/work/Wilke/tingle/phix174/output/3212022/"

# get_reports() retrieves the reports for all the simulation runs in a specified scenario
# had hardcoded case for Scenario B where there is no simulation 1
get_reports <- function(basedir, filedir, total_sims){
  report <- data.frame()
  #if(filedir == "scenario_B/") {
  #  i = 2
  #}
  #else 
  i = 1
  while (i<=total_sims) {
    sim = read_csv(paste0(basedir, filedir, "sim_", i, "_report.csv")) 
    sim$sim = i
    report <- rbind(report, sim)
    i <- i + 1
  }
  report <- report %>% 
    select(-X1)
  return(report)
}

# plot_reports() plots the RMSE v generations for all the simulations in a given scenario
plot_reports <- function(basedir, filedir, total_sims, title){
  x_scale <- c(0, 100, 200, 300, 400, 500)
  
  figure <- get_reports(basedir, filedir, total_sims) %>% 
    filter(!is.na(gen)) %>% 
    filter(gen > 0) %>% 
    filter(error < 11) %>% 
    ggplot(aes(x = gen)) +
    geom_line(aes(y = error, color = "error"), color = "pink") + 
    geom_line(aes(y = min_error, color = "min error"), color = "black") +
    facet_wrap(~sim) + 
    ggtitle(title) + 
    xlab("Generations") +
    ylab("RMSE") + 
    #ylim(c(1, 10)) + 
    scale_x_continuous(breaks = x_scale) + 
    cowplot::theme_minimal_grid(12) + 
    theme(axis.text.x = element_text(size=11, angle= -45))
  
  return(figure)
}

# logel_plot() recreates the barplot transcript abundances from Logel 2020
# sim = simulation number, gen = generation of interest within the simulation, timestamp = pinetree run
logel_plot <- function(basedir, filedir, sim, gen, timestamp){
  
  my_palette = c("#d0a34f","#75c130", "#c12c04", "#ecd575", "#b69507",
            "#e37c71", "#adcb40", "#d3ad40", "#288ede", "#a5bdfa",
            "#93c9ff")

  figure <- read_tsv(paste0(basedir, filedir,"sim_", sim, "_gen_", gen, ".tsv" )) %>% 
       mutate(time = round(time)) %>% 
       filter(time == timestamp) %>% 
       filter(!str_detect(species, '__|bound|ecoli|promoter')) %>% 
       mutate(normalized = transcript/transcript[1]) %>% 
       mutate(species = factor(species, c("gene_A", "gene_A*", "gene_B",
                                         "gene_K", "gene_C", "gene_D",
                                         "gene_E", "gene_J", "gene_F",
                                         "gene_G", "gene_H"),
                                       c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"))) %>% 
       ggplot(aes(x=species, y=normalized, fill = species)) +
       geom_bar(stat="identity", color = "black") + 
       scale_fill_manual(values = my_palette) + 
       scale_y_continuous(limits = c(0, 18), 
                          breaks = c(0,3,6,9,12,15,18)) + 
       ylab("")+ 
       cowplot::theme_cowplot(12) +
       theme(legend.position = "none") 
  
  return(figure)
}

# logeL_timeseries called logel_plot() to show how distribution changes over time (start, middle, end)
# sim = simulation number, gen = last generation of interest, timestamp = pinetree run

logel_timeseries <- function(basedir, filedir, sim, gen, timestamp) {
  
  sim_start <- logel_plot(basedir, filedir, sim, gen = 1, timestamp = 1200)+
    ylab(paste0("Sim ", sim))
  sim_middle <- logel_plot(basedir, filedir, sim, gen = gen/2, timestamp = 1200)
  sim_end <- logel_plot(basedir, filedir, sim, gen = gen, timestamp = 1200)
  
  figure <- plot_grid(sim_start, sim_middle, sim_end, 
                      label_x = "genes", label_y = "transcript abundances", 
                      ncol = 3, nrow = 1)
                     
  return(figure)
}

get_fitness <- function(file, timepoint){
  
  pt_run <- as.data.frame(read_tsv(file)) %>% 
    mutate(time = round(time, 0)) %>% 
    filter(time == timepoint) %>% 
    filter(!str_detect(species, '__|bound|ecoli|promoter')) %>% 
    mutate(species = factor(species, c("gene_A", "gene_A*", "gene_B",
                                           "gene_K", "gene_C", "gene_D",
                                           "gene_E", "gene_J", "gene_F",
                                           "gene_G", "gene_H"),
                                     c("A", "A*", "B", "K",
                                           "C", "D", "E", "J",
                                           "F", "G", "H"))) %>% 
    filter(species %in% c("B", "D", "J", "F", "G", "H")) %>% 
    select(-ribo_density, -transcript, -time) %>% 
    mutate(copies = c(60, 240, 60, 60, 12, 60))  %>% # From Leuven 2020
    mutate(stoich = protein/copies) %>%
    summarise(virions = min(stoich))
  
  return(pt_run$virions[1])
}

```

### **Objective**

Here I will analyze the results from the promoter and terminator optimization run on 3-21-2022. The optimization algorithm is the same as that used in 3-15-2022. The only difference is that there is additional spacing (about 14 extra nucleotides) between tG, and gene H. 

### **Change in Error over Generations**
These results look the best out of all the previous optimizations!! 
```{r, echo = FALSE, message = FALSE, warning=FALSE}
filedir <- ""
plot_reports(basedir, filedir, 20, "qPCR Simulations")
```

### **Final Minimum Errors**

Here, I look at the best minimal errors to see if the promoter and terminator values are converging. Note, since there are only 5 good simulations (min error < 2) at the moment it is hard to find any strong pattern. This is mostly exploratory. 

**Terminator values**

From the Logel 2020 paper, the experimental fold change analysis showed terminator strengths as follows: tJ = tG = 0.3ish and tF = 0.2ish and tH = 0.7ish. 

tJ = tG was vaguely observed in sims 3 and 7 and NOT observed in sims 4 and 5. Sims 4 and 5 had higher errors so this is fine. More (concerining?) is that it was not observed in sim 9 which has the lowest error. In sim 9, tG had a very high value....not sure why. 

**Promoter values**

From the Logel 2020 paper, pB > pD > pA, but pB and pD are fairly similar in value. In all of the best simulations, pA had the lowest value -- which is consistent with experimental observations. pD however is always stronger than pB...by 1-2 orders of magnitude. This is consistent with what I found while manually toggling the parameters...but I'm not sure why this differs from the experimental observations. 

**Outstanding Questions**

1. Terminator values are not converging convincingly enough -- plan to run for more simulations (total 30? it only took 3ish hours to run 10 sims) and generations (1000? although only incremental gains were found after 400 generations). Also use smaller step size (0.05 --> 0.01).

2. Promoter values converging but in a pattern inconsistent with experimental data (Logel 2020). I don't have a good biological reason for this -- is there some regulatory dynamic I am not modeling? 

```{r, echo = FALSE, message = FALSE, warning=FALSE}
col_order = c("sim", "gen", "pA", "pB", "pD", "tJ", "tF", "tG", "tH",  "error", "min_error")

all_sims_report <- get_reports(basedir, filedir, 20) %>% 
  group_by(sim) %>% 
  mutate(final_min_error = min(min_error)) %>% 
  filter(error == final_min_error) %>%  
  filter(row_number()==1) %>% 
  #filter(min_error < 2) %>% 
  arrange(min_error)

all_sims_report <- all_sims_report[ , col_order]

kable(all_sims_report) %>% 
  kable_styling(latex_options = "striped") %>% 
  row_spec(c(1:3), background = "#FDE1F5")

all_sims_report <- all_sims_report %>% 
  mutate(virions = get_fitness(paste0(basedir, "sim_", sim, "_gen_", gen, ".tsv"), 1200))

```

### **Final Transcript Abundances**
Sim 9 is the best in matching qPCR data from Logel 2020. 

```{r, echo = FALSE, message = FALSE, warning=FALSE}
sim9 <- logel_plot(basedir, filedir, 9, 70, 1200) + ggtitle("Sim 9")
sim1 <- logel_plot(basedir, filedir, 1, 108, 1200) +  ggtitle("Sim 1")
sim15 <- logel_plot(basedir, filedir, 15, 477, 1200) + ggtitle("Sim 15")
sim12 <- logel_plot(basedir, filedir, 12, 421, 1200) + ggtitle("Sim 12")

plot_grid(sim9, sim1, sim15, sim12, nrow = 2, ncol = 2)

```

### **Evolution of transcript abundances over generations for the best sims**
Doesn't always improve over time - which is ok since sometimes the algorithm cannot find better parameter values.
```{r, echo = FALSE, message = FALSE, warning=FALSE}
sim9_time <- logel_timeseries(basedir, filedir, 9, 500, 1200) 
sim7_time <- logel_timeseries(basedir, filedir, 7, 500, 1200) 
sim3_time <- logel_timeseries(basedir, filedir, 3, 500, 1200) 

plot_grid(sim9_time, sim7_time, sim3_time, nrow = 3) + ggtitle('Transcript Abundances Over Time')
```


### **How do parameter values change over time?**

I think the promoter values are exploring the parameter space just fine. For terminators, I want to repeat this analysis but with a smaller step size of 0.01. It needs some more fine tuning. 
```{r, echo = FALSE, message = FALSE, warning=FALSE}
get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(9,7,3)) %>% 
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim) +
  geom_line(aes(y = exp(pA))) + 
  scale_y_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  ggtitle("pA over time") +
  cowplot::theme_minimal_grid(12)

get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(9,7,3)) %>%
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim) +
  scale_y_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  geom_line(aes(y = exp(pB))) + 
  ggtitle("pB over time") +
  cowplot::theme_minimal_grid(12)


get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(9,7,3)) %>%
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim, scales="free_y") +
  geom_line(aes(y = exp(pD))) + 
  scale_y_continuous(trans = log10_trans(),
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  ggtitle("pD over time") +
  cowplot::theme_minimal_grid(12)

get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(9,7,3)) %>%
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim, scales="free_y") +
  geom_line(aes(y = tJ)) + 
  ggtitle("tJ over time") +
  cowplot::theme_minimal_grid(12)

get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(9,7,3)) %>%
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim, scales="free_y") +
  geom_line(aes(y = tF)) + 
  ggtitle("tF over time") +
  cowplot::theme_minimal_grid(12)

get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(9,7,3)) %>%
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim, scales="free_y") +
  geom_line(aes(y = tG)) + 
  ggtitle("tG over time") +
  cowplot::theme_minimal_grid(12)

get_reports(basedir, filedir, 10) %>% 
  filter(!is.na(gen)) %>% 
  filter(sim %in% c(9,7,3)) %>%
  ggplot(aes(x = gen)) +
  facet_wrap(. ~ sim, scales="free_y") +
  geom_line(aes(y = tH)) + 
  ggtitle("tH over time") +
  cowplot::theme_minimal_grid(12)
```

### **Next Steps**

Based on these results, I think I'm pretty much done with this phase of the project. Here is what I plan to do next to wrap things up:

1. **Optimize Terminator values with smaller step sizes**.
  * Instead of using steps size of 0.05, I'm going to try 0.01. Hopefully this will help it converge better. 
  * I'm also going to run for 1000 generations instead of 500. I know visually it seems like the error is minimized effectively after 400 generations...but hopefully this will provide more time/space for the terminator values to converge. 
  
  
2. **Think about why promoter values do not follow experimentally-determined trends in value**
  * Specifically, why is pD bigger than pB when the reported fold-change of pB is largest? Is there some regulatory mechanism that Pinetree is not capturing? 

### **Exploring fitness**

Here I tried using the pinetree run from the simulation/generation with the lowest error (sim 9, gen 369) to calculate viral fitness. However, NO protein H was synthesized....AND the ratios of the other proteins were too low to actually construct even 1 virion.

**Thoughts:**

1. The MOI of 5 was used experimentally...should I try dividing the cell volume by 5 to model this? Should I try a longer infection cycle? Although the typical infection length of phiX174 is like 20 mins (t = 1200). Could I register multiple genomes (ie, 5 different phiX genomes?) Maybe adding more ribosomes? 

2. ~~I'll try running pinetree with these ideas with the parameter values isolated in sim 9, gen 369.~~ EDIT: Actually sim 9, gen 369 has a very strong tG (0.96ish)...this would obviously throttle gene H expression. I'm going to use sim 7, gen 260 instead since it has a lower tG value. Also the errors between the two are only 0.05 is apart. 

```{r}
## Read pinetree run tsv into data frame. Filter and keep only the final timepoint of 1200. 
pt_run <- as.data.frame(read_tsv(file = "/stor/work/Wilke/tingle/phix174/output/3152022/sim_2_gen_2.tsv")) %>% 
  mutate(time = round(time, 0)) %>% 
  filter(time == 1200) %>% 
  filter(!str_detect(species, '__|bound|ecoli|promoter')) %>% 
  mutate(species = factor(species, c("gene_A", "gene_A*", "gene_B",
                                         "gene_K", "gene_C", "gene_D",
                                         "gene_E", "gene_J", "gene_F",
                                         "gene_G", "gene_H"),
                                   c("A", "A*", "B", "K",
                                         "C", "D", "E", "J",
                                         "F", "G", "H"))) %>% 
  filter(species %in% c("B", "D", "J", "F", "G", "H")) %>% 
  select(-transcript, -ribo_density, -time) %>% 
  mutate(copies = c(60, 240, 60, 60, 12, 60))  %>% # From Leuven 2020
  mutate(stoich = protein/copies) %>% 
  mutate(LR = min(stoich)) ## Problem - no protein H generated...

pt_run
```

```{r}

```










