{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Optimizing promoter strengths for phix-174\n",
    "In this script, I will optimize the promoter strengths for each of the three promoter sites in Phix_174. \n",
    "\n",
    "#### Main Idea\n",
    "1. Define a unique normal distribution (**~N(u,o)**) for each of the three promoters (pA, pB, pD). Let the promoter strength be **e^u_value** and step size be the st. deviation **o**. \n",
    "\n",
    "2. Define a normal distribution (**~N(0.5,?)**) to describe **lambda**. Here, **lambda** represents the value by which the step size will be scaled.\n",
    "\n",
    "3. Randomly select a promoter to optimize (pA, pB, pD). Feed promoter values into pinetree & calculate RMSE. If the new error (**NE**) is less than the old error (**OE**), then keep the promoter value. If not, increment the promoter value by the step size (add **e^o**). Keep repeating with randomly selected **o** value until **NE** < **OE**. *Stop after 5 iterations & move onto the next promoter value.*\n",
    "\n",
    "#### Key Attributes\n",
    "1. \n",
    "\n",
    "#### Terminology/Naming Conventions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Import Packages & Set Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pinetree as pt\n",
    "from sklearn.metrics import mean_squared_error \n",
    "base_dir = \"/Users/tanviingle/Documents/Wilke/phix174/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Run Pinetree\n",
    "This function runs the pinetree simulator given parameters\n",
    "\n",
    "STEPS: \n",
    "1. Feed in promoter values\n",
    "2. Run pinetree simulator\n",
    "3. Write file as gen.csv --> store as basedir/output/opt_test/gen_#.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_pt(gen, pA, pB, pD): #Start with only feeding in values for pA\n",
    "\n",
    "    print(gen, pA, pB, pD)\n",
    "    print(\"Defining PhiX-174 genome\")\n",
    "    \n",
    "    # Create host cell & genome\n",
    "    CELL_VOLUME = 1.1e-15 # from T7\n",
    "    model = pt.Model(cell_volume=CELL_VOLUME)\n",
    "    phage = pt.Genome(name=\"phix_174\", length=5386)\n",
    "    \n",
    "    # Read genomic coordinates from csv into dataframe\n",
    "    genomic_coords = pd.read_csv(base_dir + \"output/\" + \"genomic_coords.csv\")\n",
    "    print(genomic_coords.at[0, \"type\"])\n",
    "    \n",
    "    \n",
    "    # Add genomic ns (loop through ^ df); hardcode necessary strengths according to preomtimized_model\n",
    "    ## for length of genomic_coords, add elements\n",
    "    #for n in genomic_coords:{\n",
    "    n = 0\n",
    "    while(n < len(genomic_coords)):\n",
    "        \n",
    "        if genomic_coords.at[n, \"type\"] == \"gene\": \n",
    "            phage.add_gene(name= genomic_coords.at[n, \"type\"] + \"_\" + genomic_coords.at[n, \"name\"], \n",
    "                           start= genomic_coords.at[n, \"new_start\"], \n",
    "                           stop= genomic_coords.at[n, \"new_end\"],\n",
    "                           rbs_start=genomic_coords.at[n, \"new_start\"] - 15, \n",
    "                           rbs_stop=genomic_coords.at[n, \"new_start\"] - 1, rbs_strength=1e7) \n",
    "\n",
    "        elif genomic_coords.at[n, \"type\"] == \"promoter\" and genomic_coords.at[n, \"name\"] == \"A\":\n",
    "            phage.add_promoter(name= genomic_coords.at[n, \"type\"] + \"_\" + genomic_coords.at[n, \"name\"], \n",
    "                               start= genomic_coords.at[n, \"new_start\"], \n",
    "                               stop= genomic_coords.at[n, \"new_end\"],\n",
    "                               interactions={\"ecolipol\": pA})\n",
    "\n",
    "        elif genomic_coords.at[n, \"type\"] == \"promoter\" and genomic_coords.at[n, \"name\"] == \"B1\":\n",
    "            phage.add_promoter(name= genomic_coords.at[n, \"type\"] + \"_\" + genomic_coords.at[n, \"name\"], \n",
    "                               start= genomic_coords.at[n, \"new_start\"], \n",
    "                               stop= genomic_coords.at[n, \"new_end\"],\n",
    "                               interactions={\"ecolipol\": pB})\n",
    "\n",
    "        elif genomic_coords.at[n, \"type\"] == \"promoter\" and genomic_coords.at[n, \"name\"] == \"D\":\n",
    "            phage.add_promoter(name= genomic_coords.at[n, \"type\"] + \"_\" + genomic_coords.at[n, \"name\"], \n",
    "                               start= genomic_coords.at[n, \"new_start\"], \n",
    "                               stop= genomic_coords.at[n, \"new_end\"],\n",
    "                               interactions={\"ecolipol\": pD})\n",
    "\n",
    "        else:\n",
    "            print(\"ignoring pB2\")\n",
    "\n",
    "        n = n+1\n",
    "    \n",
    "    print(\"all genes and promoters added\")\n",
    "    \n",
    "    # Add terminators manually \n",
    "    phage.add_terminator(name=\"terminator_J\", start=2402, stop=2403, # Right before gene F start=2404, stop=3687,\n",
    "                       efficiency={\"ecolipol\": 0.3}) # 0.7\n",
    "    phage.add_terminator(name=\"terminator_F\", start=3796, stop=3797, # Right before gene G start=3798, stop=4325\n",
    "                     efficiency={\"ecolipol\": 0.2}) # 0.8\n",
    "    phage.add_terminator(name=\"terminator_G\", start=4332, stop=4333, # Right before gene H start=4334, stop=5320\n",
    "                     efficiency={\"ecolipol\": 0.3}) # 0.6\n",
    "    phage.add_terminator(name=\"terminator_H\", start=5321, stop=5322, # Right after gene H\n",
    "                     efficiency={\"ecolipol\": 0.7}) # 0.3\n",
    "\n",
    "    print(\"all terminators added\")\n",
    "    \n",
    "    # Register genome after promoters/terminators are added\n",
    "    model.register_genome(phage)\n",
    "    print(\"genome is registered\")\n",
    "\n",
    "    # Define interactions\n",
    "    print(\"Defining Polymerases & Interactions\")\n",
    "    # Add polymerases & species\n",
    "    model.add_polymerase(name=\"ecolipol\", speed=35, footprint=35, copy_number=0)\n",
    "    model.add_species(\"bound_ecolipol\", 2000)  # initialization, 1800\n",
    "    model.add_species(\"ecoli_genome\", 0)\n",
    "    model.add_species(\"ecoli_transcript\", 0)\n",
    "    model.add_reaction(1e7, [\"ecolipol\", \"ecoli_genome\"], [\"bound_ecolipol\"]) # 1e6\n",
    "    model.add_reaction(0.04, [\"bound_ecolipol\"], [\"ecolipol\", \"ecoli_genome\", \"ecoli_transcript\"])\n",
    "    model.add_ribosome(10, 30, 100)\n",
    "    model.add_species(\"bound_ribosome\", 100)\n",
    "    model.seed(34)\n",
    "    \n",
    "    # Run simulation\n",
    "    print(\"running simulation\")\n",
    "    model.simulate(time_limit=2400, time_step=5, output= base_dir + \"output/opt_test/gen_\"+str(gen)+\".tsv\") # TODO change limit\n",
    "    print(\"Simulation successful!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 200787.01532646132 1672784.230427204 1342440.7898530285\n",
      "Defining PhiX-174 genome\n",
      "gene\n",
      "ignoring pB2\n",
      "all genes and promoters added\n",
      "all terminators added\n",
      "genome is registered\n",
      "Defining Polymerases & Interactions\n",
      "running simulation\n",
      "Simulation successful!\n"
     ]
    }
   ],
   "source": [
    "run_pt(gen = 1, \n",
    "       pA = np.exp(12.21), \n",
    "       pB = np.exp(14.33), \n",
    "       pD = np.exp(14.11))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Calculate Error\n",
    "This script compares a pinetree run to the experimental data\n",
    "STEPS:\n",
    "1. Read pinetree run file\n",
    "2. Use RMSE to calculate error \n",
    "3. Return error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_error(file):\n",
    "    sim = pd.read_csv(file, sep = \"\\t\")\n",
    "    sim = sim.round({'time': 0})\n",
    "    sim = sim[sim['time'] == 2400.0]\n",
    "    sim = sim[sim.species.str.match(\"gene_\")]\n",
    "    sim[\"norm\"] = sim['transcript']/(sim.iloc[0][\"transcript\"])\n",
    "    sim[\"exp\"] = [1, 1, 6, 6, 17, 17, 11, 5, 1, 17, 6]\n",
    "    error = mean_squared_error(sim.exp, sim.norm, squared = False)\n",
    "    return(error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Define Normal Distributions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 [8204.11269147] [2366.72151246] [34327.80847346]\n",
      "Defining PhiX-174 genome\n",
      "gene\n",
      "ignoring pB2\n",
      "all genes and promoters added\n",
      "all terminators added\n",
      "genome is registered\n",
      "Defining Polymerases & Interactions\n",
      "running simulation\n",
      "Simulation successful!\n",
      "good starting values\n",
      "\n",
      "\n",
      "pB\n",
      "[2366.72151246]\n",
      "increment pB\n",
      "0 [8204.11269147] [2369.72151246] [34327.80847346]\n",
      "Defining PhiX-174 genome\n",
      "gene\n",
      "ignoring pB2\n",
      "all genes and promoters added\n",
      "all terminators added\n",
      "genome is registered\n",
      "Defining Polymerases & Interactions\n",
      "running simulation\n",
      "Simulation successful!\n",
      "100.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gen</th>\n",
       "      <th>pA</th>\n",
       "      <th>pB</th>\n",
       "      <th>pD</th>\n",
       "      <th>error</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NA</td>\n",
       "      <td>[8204.112691470798]</td>\n",
       "      <td>[2366.7215124589356]</td>\n",
       "      <td>[34327.80847346141]</td>\n",
       "      <td>1.000000e+10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>[8204.112691470798]</td>\n",
       "      <td>[2369.7215124589356]</td>\n",
       "      <td>[34327.80847346141]</td>\n",
       "      <td>1.000000e+10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gen                   pA                    pB                   pD  \\\n",
       "0  NA  [8204.112691470798]  [2366.7215124589356]  [34327.80847346141]   \n",
       "1   0  [8204.112691470798]  [2369.7215124589356]  [34327.80847346141]   \n",
       "\n",
       "          error  \n",
       "0  1.000000e+10  \n",
       "1  1.000000e+10  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "pB\n",
      "[2.68811714e+43]\n",
      "increment pB\n",
      "1 [8204.11269147] [2.68811714e+43] [34327.80847346]\n",
      "Defining PhiX-174 genome\n",
      "gene\n",
      "ignoring pB2\n",
      "all genes and promoters added\n",
      "all terminators added\n",
      "genome is registered\n",
      "Defining Polymerases & Interactions\n",
      "running simulation\n"
     ]
    },
    {
     "ename": "RuntimeError",
     "evalue": "Gillespie: Propensity of system is 0. No reactions will execute.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-29-4be74b33da30>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     62\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"increment pB\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     63\u001b[0m         \u001b[0mpB\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpB\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0moB\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 64\u001b[0;31m         \u001b[0mrun_pt\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgen\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgen\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpA\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpA\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpB\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpB\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpD\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpD\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     65\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     66\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mpX\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"pD\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-21-0a22b337ecdb>\u001b[0m in \u001b[0;36mrun_pt\u001b[0;34m(gen, pA, pB, pD)\u001b[0m\n\u001b[1;32m     83\u001b[0m     \u001b[0;31m# Run simulation\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     84\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"running simulation\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 85\u001b[0;31m     \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msimulate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtime_limit\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m2400\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtime_step\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m5\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moutput\u001b[0m\u001b[0;34m=\u001b[0m \u001b[0mbase_dir\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m\"output/opt_test/gen_\"\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\".tsv\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# TODO change limit\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     86\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Simulation successful!\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mRuntimeError\u001b[0m: Gillespie: Propensity of system is 0. No reactions will execute."
     ]
    }
   ],
   "source": [
    "gen = 0\n",
    "error = 1e10\n",
    "step = 1e2 # some error with reassigning this value, scope? \n",
    "\n",
    "uA = 10\n",
    "oA = 2\n",
    "pA = np.exp(np.random.normal(uA, oA, 1))\n",
    "\n",
    "uB = 12\n",
    "oB = 3\n",
    "pB = np.exp(np.random.normal(uB, oB, 1))\n",
    "\n",
    "uD = 12\n",
    "oD = 2\n",
    "pD = np.exp(np.random.normal(uD, oD, 1))\n",
    "\n",
    "j = 0\n",
    "\n",
    "params = [\"pA\", \"pB\", \"pD\"]\n",
    "report_df = pd.DataFrame(columns = ['gen', 'pA', 'pB', 'pD', 'error'])\n",
    "\n",
    "first_run = {'gen': \"NA\", 'pA': pA, 'pB': pB, 'pD': pD,'error': error} \n",
    "report_df = report_df.append(first_run, ignore_index = True)\n",
    "\n",
    "old_error = error\n",
    "new_error = error\n",
    "\n",
    "no_change = 0\n",
    "\n",
    "while (old_error >= new_error and no_change <= 5):\n",
    "    # STEP 1: Randomly select promoter to optimize; if gen == 10, quit. \n",
    "    if (gen == 10):\n",
    "        break   \n",
    "    \n",
    "    while (gen == 0 and j == 0):\n",
    "        try:\n",
    "            run_pt(gen, pA, pB, pD)\n",
    "            j = 1\n",
    "            print (\"good starting values\")\n",
    "            print (f\"\\n\")\n",
    "        except:\n",
    "            print(\"poor starting values chosen, retrying\")\n",
    "            pA = np.exp(np.random.normal(uA, oA, 1))\n",
    "            pB = np.exp(np.random.normal(uB, oB, 1))\n",
    "            pD = np.exp(np.random.normal(uD, oD, 1))\n",
    "            j = 0 \n",
    "            \n",
    "    pX = np.random.choice(params)\n",
    "    \n",
    "    print(pX)\n",
    "          \n",
    "    # STEP 2: Run pinetree with selected promoter & value\n",
    "    # Increment pX by oX\n",
    "    if (pX == \"pA\"):\n",
    "        print (pA)\n",
    "        print(\"increment pA\")\n",
    "        pA = pA + oA \n",
    "        run_pt(gen = gen, pA = pA, pB = pB, pD = pD)\n",
    "            \n",
    "    if (pX == \"pB\"):\n",
    "        print (pB)\n",
    "        print(\"increment pB\")\n",
    "        pB = pB + oB\n",
    "        run_pt(gen = gen, pA = pA, pB = pB, pD = pD)\n",
    "        \n",
    "    if (pX == \"pD\"):\n",
    "        print (pD)\n",
    "        print(\"increment pD\")\n",
    "        pD = pD + oD\n",
    "        run_pt(gen = gen, pA = pA, pB = pB, pD = pD)\n",
    "    \n",
    "    # STEP 3: Calculate Error\n",
    "    old_error = new_error\n",
    "    new_error = get_error(file = base_dir + \"output/opt_test/gen_\" +str(gen)+\".tsv\")\n",
    "    \n",
    "    # STEP 4: Compare Old Error to New Error; \n",
    "    if (new_error > old_error):\n",
    "        no_change = no_change + 1\n",
    "        print(\"new_error = \" + str(new_error) + \", old_error = \" + str(old_error) + \"\\n\")\n",
    "        new_error = old_error\n",
    "        old_error = report_df.at[gen-1, \"error\"]\n",
    "        if (pX == \"pA\"):\n",
    "            step = oA*np.random.normal(0.5, 0.2, 1)\n",
    "            pA = report_df.at[gen-1, \"pA\"] + np.exp(step)\n",
    "            continue\n",
    "        if (pX == \"pB\"):\n",
    "            step = oB*np.random.normal(0.5, 0.2, 1)\n",
    "            pB = report_df.at[gen-1, \"pB\"] + np.exp(step)\n",
    "            continue\n",
    "        if (pX == \"pD\"):\n",
    "            step = oD*np.random.normal(0.5, 0.2, 1)\n",
    "            pD = report_df.at[gen-1, \"pD\"] + np.exp(step)\n",
    "            continue\n",
    "    \n",
    "    print (step)    \n",
    "    new_run = {'gen': gen, 'pA': pA, 'pB': pB, 'pD': pD,'error': error} \n",
    "    report_df = report_df.append(new_run, ignore_index = True)\n",
    "    if (pX == \"pA\"):\n",
    "        pA = pA + np.exp(step)\n",
    "    if (pX == \"pB\"):\n",
    "        pB = pB + np.exp(step)\n",
    "    if (pX == \"pD\"):\n",
    "        pD = pD + np.exp(step)\n",
    "   \n",
    "    gen = gen + 1\n",
    "    display(report_df)\n",
    "    print(f\"\\n\")\n",
    "        \n",
    "#display(report_df)\n",
    "report_df.to_csv(base_dir + \"output/opt_test/relative_report_test.csv\")\n",
    "display(report_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 200000.0\n",
      "Defining PhiX-174 genome\n",
      "gene\n",
      "ignoring pB2\n",
      "all genes and promoters added\n",
      "all terminators added\n",
      "genome is registered\n",
      "Defining Polymerases & Interactions\n",
      "running simulation\n",
      "Simulation successful!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "4.703955779725772"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pA = 2e5\n",
    "pB = 62.5*pA\n",
    "pD = 52.5*pA\n",
    "run_pt(gen = 0, pA = pA, pB = pB, pD = pD)\n",
    "#run_pt(gen = 0, pA = 2e5, pB = 5e6, pD = 2e10) --> RMSE = 3.58\n",
    "#run_pt(gen = 0, pA = 2e3, pB = 5e4, pD = 2e5) --> RMSE = 36.5\n",
    "gen = 0\n",
    "get_error(file = base_dir + \"output/opt_test/gen_\" +str(gen)+\".tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "gen = 0\n",
    "error = 1e10\n",
    "step = np.exp(oA)\n",
    "report_df = pd.DataFrame(columns = ['gen', 'pA', 'error'])\n",
    "\n",
    "first_run = {'gen': \"NA\", 'pA': pA, 'error': error} \n",
    "report_df = report_df.append(first_run, ignore_index = True)\n",
    "   \n",
    "old_error = error\n",
    "new_error = error\n",
    "\n",
    "no_change = 0\n",
    "\n",
    "# Introduce while loop conditional \n",
    "while (old_error >= new_error and no_change <= 5):\n",
    "    run_pt(gen, pA) # output = file\n",
    "    old_error = new_error\n",
    "    new_error = get_error(file = base_dir + \"output/opt_test/gen_\" +str(gen)+\".tsv\") # output error\n",
    "    if(new_error > old_error):\n",
    "        no_change = no_change + 1\n",
    "        print(\"new_error = \" + str(new_error) + \", old_error = \" + str(old_error) + \"\\n\")\n",
    "        # this means we've stepped too far\n",
    "        # back up and use smaller step!\n",
    "        step = oA*np.random.normal(0.5, 0.2, 1)\n",
    "        new_error = old_error\n",
    "        old_error = report_df.at[gen-1, \"error\"]\n",
    "        pA = report_df.at[gen-1, \"pA\"] + np.exp(step)\n",
    "        continue\n",
    "    new_run = {'gen': gen, 'pA': pA, 'error': new_error} \n",
    "    report_df = report_df.append(new_run, ignore_index = True)\n",
    "    if gen == 6:\n",
    "        break\n",
    "    pA = pA + np.exp(step)\n",
    "    gen = gen + 1\n",
    "    #display(report_df)\n",
    "    print(f\"\\n\")\n",
    "\n",
    "#display(report_df)\n",
    "report_df.to_csv(base_dir + \"output/opt_test/relative_report_test.csv\")\n",
    "display(report_df)\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "147266.6252405527\n",
      "0 147266.6252405527\n",
      "Defining PhiX-174 genome\n",
      "gene\n",
      "ignoring pB2\n",
      "all genes and promoters added\n",
      "all terminators added\n",
      "genome is registered\n",
      "Defining Polymerases & Interactions\n",
      "running simulation\n",
      "Simulation successful!\n"
     ]
    }
   ],
   "source": [
    "''''\n",
    "#### Weird Pinetree Overlap Errors\n",
    "Some promoter values cause an overlap error in the simulation.\n",
    "Here I test using try/except to find promoter values to test \n",
    "that do not result in this error \n",
    "j = 0\n",
    "k = 0\n",
    "pA = np.exp(11.9)\n",
    "\n",
    "while (j==0 and k < 100):\n",
    "    print(k)\n",
    "    try:\n",
    "        print(pA)\n",
    "        run_pt(gen = 0, pA = pA, pB = 5e6, pD = 2e5)\n",
    "        j = 1\n",
    "    except:\n",
    "        print(\"Overlap issue\")\n",
    "        print(f\"\\n\")\n",
    "        j = 0\n",
    "        pA = pA + np.exp(6*np.random.normal(0.5, 0.2, 1))\n",
    "        k = k + 1\n",
    "        \n",
    "        \n",
    "uA = 12.21\n",
    "oA = 2\n",
    "pA = np.exp(np.random.normal(uA, oA, 1))\n",
    "\n",
    "uB = 17.7\n",
    "oB = 3\n",
    "pB = np.exp(np.random.normal(uB, oB, 1))\n",
    "\n",
    "uD = 14.5\n",
    "oD = 2\n",
    "pD = np.exp(np.random.normal(uD, oD, 1))\n",
    "''''''"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
