{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Optimizing promoter strengths for phix-174\n",
    "In this script, I will optimize the promoter strengths for each of the three promoter sites in Phix_174. \n",
    "\n",
    "#### Main Idea\n",
    "1. Define a unique normal distribution (**~N(u,o)**) for each of the three promoters (pA, pB, pD). Let the promoter strength be **e^u_value** and step size be the st. deviation **o**. \n",
    "\n",
    "2. Define a normal distribution (**~N(0.5,?)**) to describe **lambda**. Here, **lambda** represents the value by which the step size will be scaled.\n",
    "\n",
    "3. Randomly select a promoter to optimize (pA, pB, pD). Feed promoter values into pinetree & calculate RMSE. If the new error (**NE**) is less than the old error (**OE**), then keep the promoter value. If not, increment the promoter value by the step size (add **e^o**). Keep repeating with randomly selected **o** value until **NE** < **OE**. *Stop after 5 iterations & move onto the next promoter value.*\n",
    "\n",
    "#### Key Attributes\n",
    "1. \n",
    "\n",
    "#### Terminology/Naming Conventions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Import Packages & Set Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pinetree as pt\n",
    "from sklearn.metrics import mean_squared_error \n",
    "base_dir = \"/Users/tanviingle/Documents/Wilke/phix174/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Run Pinetree\n",
    "This function runs the pinetree simulator given parameters\n",
    "\n",
    "STEPS: \n",
    "1. Feed in promoter values\n",
    "2. Run pinetree simulator\n",
    "3. Write file as gen.csv --> store as basedir/output/opt_test/gen_#.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_pt(gen, pA): #Start with only feeding in values for pA\n",
    "\n",
    "    print(gen, pA)\n",
    "    print(\"Defining PhiX-174 genome\")\n",
    "    \n",
    "    # Create host cell & genome\n",
    "    CELL_VOLUME = 1.1e-15 # from T7\n",
    "    model = pt.Model(cell_volume=CELL_VOLUME)\n",
    "    phage = pt.Genome(name=\"phix_174\", length=5386)\n",
    "    \n",
    "    # Read genomic coordinates from csv into dataframe\n",
    "    genomic_coords = pd.read_csv(base_dir + \"output/\" + \"genomic_coords.csv\")\n",
    "    print(genomic_coords.at[0, \"type\"])\n",
    "    \n",
    "    \n",
    "    # Add genomic ns (loop through ^ df); hardcode necessary strengths according to preomtimized_model\n",
    "    ## for length of genomic_coords, add elements\n",
    "    #for n in genomic_coords:{\n",
    "    n = 0\n",
    "    while(n < len(genomic_coords)):\n",
    "        \n",
    "        if genomic_coords.at[n, \"type\"] == \"gene\": \n",
    "            phage.add_gene(name= genomic_coords.at[n, \"type\"] + \"_\" + genomic_coords.at[n, \"name\"], \n",
    "                           start= genomic_coords.at[n, \"new_start\"], \n",
    "                           stop= genomic_coords.at[n, \"new_end\"],\n",
    "                           rbs_start=genomic_coords.at[n, \"new_start\"], \n",
    "                           rbs_stop=genomic_coords.at[n, \"new_start\"] + 15, rbs_strength=1e7) \n",
    "\n",
    "        elif genomic_coords.at[n, \"type\"] == \"promoter\" and genomic_coords.at[n, \"name\"] == \"A\":\n",
    "            phage.add_promoter(name= genomic_coords.at[n, \"type\"] + \"_\" + genomic_coords.at[n, \"name\"], \n",
    "                               start= genomic_coords.at[n, \"new_start\"], \n",
    "                               stop= genomic_coords.at[n, \"new_end\"],\n",
    "                               interactions={\"ecolipol\": pA})\n",
    "\n",
    "        elif genomic_coords.at[n, \"type\"] == \"promoter\" and genomic_coords.at[n, \"name\"] == \"B1\":\n",
    "            phage.add_promoter(name= genomic_coords.at[n, \"type\"] + \"_\" + genomic_coords.at[n, \"name\"], \n",
    "                               start= genomic_coords.at[n, \"new_start\"], \n",
    "                               stop= genomic_coords.at[n, \"new_end\"],\n",
    "                               interactions={\"ecolipol\": 5e7})\n",
    "\n",
    "        elif genomic_coords.at[n, \"type\"] == \"promoter\" and genomic_coords.at[n, \"name\"] == \"D\":\n",
    "            phage.add_promoter(name= genomic_coords.at[n, \"type\"] + \"_\" + genomic_coords.at[n, \"name\"], \n",
    "                               start= genomic_coords.at[n, \"new_start\"], \n",
    "                               stop= genomic_coords.at[n, \"new_end\"],\n",
    "                               interactions={\"ecolipol\": 2e6})\n",
    "\n",
    "        else:\n",
    "            print(\"ignoring pB2\")\n",
    "\n",
    "        n = n+1\n",
    "    \n",
    "    print(\"all genes and promoters added\")\n",
    "    \n",
    "    # Add terminators manually \n",
    "    phage.add_terminator(name=\"terminator_J\", start=2402, stop=2403, # Right before gene F start=2404, stop=3687,\n",
    "                       efficiency={\"ecolipol\": 0.7}) # 0.7\n",
    "    phage.add_terminator(name=\"terminator_F\", start=3796, stop=3797, # Right before gene G start=3798, stop=4325\n",
    "                     efficiency={\"ecolipol\": 0.8}) # 0.8\n",
    "    phage.add_terminator(name=\"terminator_G\", start=4332, stop=4333, # Right before gene H start=4334, stop=5320\n",
    "                     efficiency={\"ecolipol\": 0.6}) # 0.6\n",
    "    phage.add_terminator(name=\"terminator_H\", start=5321, stop=5322, # Right after gene H\n",
    "                     efficiency={\"ecolipol\": 0.3}) # 0.3\n",
    "\n",
    "    print(\"all terminators added\")\n",
    "    \n",
    "    # Register genome after promoters/terminators are added\n",
    "    model.register_genome(phage)\n",
    "    print(\"genome is registered\")\n",
    "\n",
    "    # Define interactions\n",
    "    print(\"Defining Polymerases & Interactions\")\n",
    "    # Add polymerases & species\n",
    "    model.add_polymerase(name=\"ecolipol\", speed=35, footprint=35, copy_number=0)\n",
    "    model.add_species(\"bound_ecolipol\", 1800)  # initialization\n",
    "    model.add_species(\"ecoli_genome\", 0)\n",
    "    model.add_species(\"ecoli_transcript\", 0)\n",
    "    model.add_reaction(1e6, [\"ecolipol\", \"ecoli_genome\"], [\"bound_ecolipol\"]) # 1e7\n",
    "    model.add_reaction(0.04, [\"bound_ecolipol\"], [\"ecolipol\", \"ecoli_genome\", \"ecoli_transcript\"])\n",
    "    model.add_ribosome(10, 30, 100)\n",
    "    model.add_species(\"bound_ribosome\", 100)\n",
    "    model.seed(34)\n",
    "    \n",
    "    # Run simulation\n",
    "    print(\"running simulation\")\n",
    "    model.simulate(time_limit=1200, time_step=5, output= base_dir + \"output/opt_test/gen_\"+str(gen)+\".tsv\") # TODO change limit\n",
    "    print(\"Simulation successful!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Calculate Error\n",
    "This script compares a pinetree run to the experimental data\n",
    "STEPS:\n",
    "1. Read pinetree run file\n",
    "2. Use RMSE to calculate error \n",
    "3. Return error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_error(file):\n",
    "    sim = pd.read_csv(file, sep = \"\\t\")\n",
    "    sim = sim.round({'time': 0})\n",
    "    sim = sim[sim['time'] == 1200.0]\n",
    "    sim = sim[sim.species.str.match(\"gene_\")]\n",
    "    sim[\"norm\"] = sim['transcript']/(sim.iloc[0][\"transcript\"])\n",
    "    sim[\"exp\"] = [1, 1, 6, 6, 17, 17, 11, 5, 1, 17, 6]\n",
    "    error = mean_squared_error(sim.exp, sim.norm, squared = False)\n",
    "    return(error)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Define Normal Distributions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "uA = 12.21\n",
    "oA = 2\n",
    "pA = np.exp(np.random.normal(uA, oA, 1))\n",
    "\n",
    "uB = 17.7\n",
    "oB = 3\n",
    "pB = np.exp(np.random.normal(uB, oB, 1))\n",
    "\n",
    "uD = 14.5\n",
    "oD = 2\n",
    "pD = np.exp(np.random.normal(uD, oD, 1))\n",
    "\n",
    "l = np.random.normal(0.5, 0.2, 1) # add code to resample if l < 0 or l > 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gen = 0\n",
    "error = 1e10\n",
    "step = np.exp(oA)\n",
    "report_df = pd.DataFrame(columns = ['gen', 'pA', 'error'])\n",
    "\n",
    "first_run = {'gen': \"NA\", 'pA': pA, 'error': error} \n",
    "report_df = report_df.append(first_run, ignore_index = True)\n",
    "\n",
    "old_error = error\n",
    "new_error = error\n",
    "\n",
    "no_change = 0\n",
    "\n",
    "# Introduce while loop conditional \n",
    "while (old_error >= new_error and no_change <= 5):\n",
    "    run_pt(gen, pA) # output = file\n",
    "    old_error = new_error\n",
    "    new_error = get_error(file = base_dir + \"output/opt_test/gen_\" +str(gen)+\".tsv\") # output error\n",
    "    if(new_error > old_error):\n",
    "        no_change = no_change + 1\n",
    "        print(\"new_error = \" + str(new_error) + \", old_error = \" + str(old_error) + \"\\n\")\n",
    "        # this means we've stepped too far\n",
    "        # back up and use smaller step!\n",
    "        step = oA*np.random.normal(0.5, 0.2, 1)\n",
    "        new_error = old_error\n",
    "        old_error = report_df.at[gen-1, \"error\"]\n",
    "        pA = report_df.at[gen-1, \"pA\"] + np.exp(step)\n",
    "        continue\n",
    "    new_run = {'gen': gen, 'pA': pA, 'error': new_error} \n",
    "    report_df = report_df.append(new_run, ignore_index = True)\n",
    "    if gen == 6:\n",
    "        break\n",
    "    pA = pA + np.exp(step)\n",
    "    gen = gen + 1\n",
    "    #display(report_df)\n",
    "    print(f\"\\n\")\n",
    "\n",
    "#display(report_df)\n",
    "report_df.to_csv(base_dir + \"output/opt_test/relative_report_test.csv\")\n",
    "display(report_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# randomly select item from a list of 3\n",
    "params = [\"pA\", \"pB\", \"pD\"]\n",
    "print(np.random.choice(params))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gen = 0\n",
    "error = 1e10\n",
    "step = np.exp(oA)\n",
    "report_df = pd.DataFrame(columns = ['gen', 'pA', 'pB', 'pD', 'error'])\n",
    "\n",
    "first_run = {'gen': \"NA\", 'pA': pA, 'pB': pB, 'pD': pD,'error': error} \n",
    "report_df = report_df.append(first_run, ignore_index = True)\n",
    "\n",
    "old_error = error\n",
    "new_error = error\n",
    "\n",
    "no_change = 0\n",
    "\n",
    "# Introduce while loop conditional \n",
    "while (old_error >= new_error and no_change <= 5):\n",
    "    run_pt(gen, pA) # output = file\n",
    "    old_error = new_error\n",
    "    new_error = get_error(file = base_dir + \"output/opt_test/gen_\" +str(gen)+\".tsv\") # output error\n",
    "    if(new_error > old_error):\n",
    "        no_change = no_change + 1\n",
    "        print(\"new_error = \" + str(new_error) + \", old_error = \" + str(old_error) + \"\\n\")\n",
    "        # this means we've stepped too far\n",
    "        # back up and use smaller step!\n",
    "        step = oA*np.random.normal(0.5, 0.2, 1)\n",
    "        new_error = old_error\n",
    "        old_error = report_df.at[gen-1, \"error\"]\n",
    "        pA = report_df.at[gen-1, \"pA\"] + np.exp(step)\n",
    "        continue\n",
    "    new_run = {'gen': gen, 'pA': pA, 'error': new_error} \n",
    "    report_df = report_df.append(new_run, ignore_index = True)\n",
    "    if gen == 6:\n",
    "        break\n",
    "    pA = pA + np.exp(step)\n",
    "    gen = gen + 1\n",
    "    #display(report_df)\n",
    "    print(f\"\\n\")\n",
    "\n",
    "#display(report_df)\n",
    "report_df.to_csv(base_dir + \"output/opt_test/relative_report_test.csv\")\n",
    "display(report_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
